<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.17"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>vg: vg::HaplotypeIndexer Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">vg
   </div>
   <div id="projectbrief">tools for working with variation graphs</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.17 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacevg.html">vg</a></li><li class="navelem"><a class="el" href="classvg_1_1HaplotypeIndexer.html">HaplotypeIndexer</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="classvg_1_1HaplotypeIndexer-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">vg::HaplotypeIndexer Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p><code>#include &lt;haplotype_indexer.hpp&gt;</code></p>
<div class="dynheader">
Inheritance diagram for vg::HaplotypeIndexer:</div>
<div class="dyncontent">
 <div class="center">
  <img src="classvg_1_1HaplotypeIndexer.png" usemap="#vg::HaplotypeIndexer_map" alt=""/>
  <map id="vg::HaplotypeIndexer_map" name="vg::HaplotypeIndexer_map">
<area href="classvg_1_1Progressive.html" alt="vg::Progressive" shape="rect" coords="0,0,131,24"/>
  </map>
</div></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:ae0b8cbfe64f32152ceefa57daf9ece72"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#ae0b8cbfe64f32152ceefa57daf9ece72">HaplotypeIndexer</a> ()</td></tr>
<tr class="memdesc:ae0b8cbfe64f32152ceefa57daf9ece72"><td class="mdescLeft">&#160;</td><td class="mdescRight">Perform initialization of backing libraries.  <a href="classvg_1_1HaplotypeIndexer.html#ae0b8cbfe64f32152ceefa57daf9ece72">More...</a><br /></td></tr>
<tr class="separator:ae0b8cbfe64f32152ceefa57daf9ece72"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad3eda4acd04b8bae303b4604410ece41"><td class="memItemLeft" align="right" valign="top">std::vector&lt; std::string &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#ad3eda4acd04b8bae303b4604410ece41">parse_vcf</a> (const std::string &amp;filename, const <a class="el" href="classhandlegraph_1_1PathHandleGraph.html">PathHandleGraph</a> &amp;graph, const std::string &amp;job_name=&quot;GBWT&quot;) const</td></tr>
<tr class="separator:ad3eda4acd04b8bae303b4604410ece41"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad81dbd24bba40aaa14536d79754d387b"><td class="memItemLeft" align="right" valign="top">std::vector&lt; std::string &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#ad81dbd24bba40aaa14536d79754d387b">parse_vcf</a> (const std::string &amp;filename, const <a class="el" href="classhandlegraph_1_1PathHandleGraph.html">PathHandleGraph</a> &amp;graph, const std::vector&lt; <a class="el" href="structhandlegraph_1_1path__handle__t.html">path_handle_t</a> &gt; &amp;paths, const std::string &amp;job_name=&quot;GBWT&quot;) const</td></tr>
<tr class="separator:ad81dbd24bba40aaa14536d79754d387b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0389ee96c5ce8a6d9c158fda39aca9d9"><td class="memItemLeft" align="right" valign="top">std::unique_ptr&lt; gbwt::DynamicGBWT &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#a0389ee96c5ce8a6d9c158fda39aca9d9">build_gbwt</a> (const std::vector&lt; std::string &gt; &amp;vcf_parse_files, const std::string &amp;job_name=&quot;GBWT&quot;, const <a class="el" href="classhandlegraph_1_1PathHandleGraph.html">PathHandleGraph</a> *graph=nullptr, const std::unordered_set&lt; <a class="el" href="structhandlegraph_1_1path__handle__t.html">path_handle_t</a> &gt; *paths=nullptr, bool skip_unvisited_paths=false) const</td></tr>
<tr class="separator:a0389ee96c5ce8a6d9c158fda39aca9d9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae92d109c2e7baf0b7ddded8df80a6f70"><td class="memItemLeft" align="right" valign="top">std::unique_ptr&lt; gbwt::DynamicGBWT &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#ae92d109c2e7baf0b7ddded8df80a6f70">build_gbwt</a> (const <a class="el" href="classhandlegraph_1_1PathHandleGraph.html">PathHandleGraph</a> &amp;graph) const</td></tr>
<tr class="separator:ae92d109c2e7baf0b7ddded8df80a6f70"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a34fc36c5ba77f4cedd32073dda118b84"><td class="memItemLeft" align="right" valign="top">std::unique_ptr&lt; gbwt::GBWT &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#a34fc36c5ba77f4cedd32073dda118b84">build_gbwt</a> (const <a class="el" href="classhandlegraph_1_1HandleGraph.html">HandleGraph</a> &amp;graph, const std::vector&lt; std::string &gt; &amp;aln_filenames, const std::string &amp;aln_format, size_t parallel_jobs=1) const</td></tr>
<tr class="separator:a34fc36c5ba77f4cedd32073dda118b84"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="inherit_header pub_methods_classvg_1_1Progressive"><td colspan="2" onclick="javascript:toggleInherit('pub_methods_classvg_1_1Progressive')"><img src="closed.png" alt="-"/>&#160;Public Member Functions inherited from <a class="el" href="classvg_1_1Progressive.html">vg::Progressive</a></td></tr>
<tr class="memitem:a4eb5050eb36ed519cbfa398748ef183b inherit pub_methods_classvg_1_1Progressive"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Progressive.html#a4eb5050eb36ed519cbfa398748ef183b">preload_progress</a> (const string &amp;message)</td></tr>
<tr class="separator:a4eb5050eb36ed519cbfa398748ef183b inherit pub_methods_classvg_1_1Progressive"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1cce3f715516ed2bbb4ce3a312b01e02 inherit pub_methods_classvg_1_1Progressive"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Progressive.html#a1cce3f715516ed2bbb4ce3a312b01e02">create_progress</a> (const string &amp;message, long count)</td></tr>
<tr class="separator:a1cce3f715516ed2bbb4ce3a312b01e02 inherit pub_methods_classvg_1_1Progressive"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a75d4e92a36a277be5bc0cee5d01f5ca1 inherit pub_methods_classvg_1_1Progressive"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Progressive.html#a75d4e92a36a277be5bc0cee5d01f5ca1">create_progress</a> (long count)</td></tr>
<tr class="separator:a75d4e92a36a277be5bc0cee5d01f5ca1 inherit pub_methods_classvg_1_1Progressive"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0d64e6aa69bc094c163f4827602a302c inherit pub_methods_classvg_1_1Progressive"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Progressive.html#a0d64e6aa69bc094c163f4827602a302c">update_progress</a> (long i)</td></tr>
<tr class="separator:a0d64e6aa69bc094c163f4827602a302c inherit pub_methods_classvg_1_1Progressive"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab1de11917caf25974a02e2f5c3946280 inherit pub_methods_classvg_1_1Progressive"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Progressive.html#ab1de11917caf25974a02e2f5c3946280">increment_progress</a> ()</td></tr>
<tr class="separator:ab1de11917caf25974a02e2f5c3946280 inherit pub_methods_classvg_1_1Progressive"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaab8c6ae92870077bb59f86a15c8d18f inherit pub_methods_classvg_1_1Progressive"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Progressive.html#aaab8c6ae92870077bb59f86a15c8d18f">destroy_progress</a> (void)</td></tr>
<tr class="separator:aaab8c6ae92870077bb59f86a15c8d18f inherit pub_methods_classvg_1_1Progressive"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:afb3592792a7d781cb6e88fe7949b7bfe"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#afb3592792a7d781cb6e88fe7949b7bfe">warn_on_missing_variants</a> = true</td></tr>
<tr class="memdesc:afb3592792a7d781cb6e88fe7949b7bfe"><td class="mdescLeft">&#160;</td><td class="mdescRight">Print a warning if variants in the VCF can't be found in the graph.  <a href="classvg_1_1HaplotypeIndexer.html#afb3592792a7d781cb6e88fe7949b7bfe">More...</a><br /></td></tr>
<tr class="separator:afb3592792a7d781cb6e88fe7949b7bfe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a512203218e3b78afa0fc5c0e594e6516"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#a512203218e3b78afa0fc5c0e594e6516">max_missing_variant_warnings</a> = 10</td></tr>
<tr class="memdesc:a512203218e3b78afa0fc5c0e594e6516"><td class="mdescLeft">&#160;</td><td class="mdescRight">Only report up to this many of them.  <a href="classvg_1_1HaplotypeIndexer.html#a512203218e3b78afa0fc5c0e594e6516">More...</a><br /></td></tr>
<tr class="separator:a512203218e3b78afa0fc5c0e594e6516"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5ff016a7fdf425ec4f3d5408d09b951f"><td class="memItemLeft" align="right" valign="top">std::map&lt; std::string, std::string &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#a5ff016a7fdf425ec4f3d5408d09b951f">path_to_vcf</a></td></tr>
<tr class="separator:a5ff016a7fdf425ec4f3d5408d09b951f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8061130697dc26811303c4b67b081245"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#a8061130697dc26811303c4b67b081245">rename_variants</a> = true</td></tr>
<tr class="separator:a8061130697dc26811303c4b67b081245"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa818eebd7cffc81773e8bea65bd398cb"><td class="memItemLeft" align="right" valign="top">std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#aa818eebd7cffc81773e8bea65bd398cb">batch_file_prefix</a> = &quot;&quot;</td></tr>
<tr class="separator:aa818eebd7cffc81773e8bea65bd398cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a881e755025092748b18f037fb1078e65"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#a881e755025092748b18f037fb1078e65">phase_homozygous</a> = true</td></tr>
<tr class="memdesc:a881e755025092748b18f037fb1078e65"><td class="mdescLeft">&#160;</td><td class="mdescRight">Phase homozygous unphased variants.  <a href="classvg_1_1HaplotypeIndexer.html#a881e755025092748b18f037fb1078e65">More...</a><br /></td></tr>
<tr class="separator:a881e755025092748b18f037fb1078e65"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acb44a055288c0a9debed5662e461959e"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#acb44a055288c0a9debed5662e461959e">force_phasing</a> = false</td></tr>
<tr class="memdesc:acb44a055288c0a9debed5662e461959e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Arbitrarily phase all unphased variants.  <a href="classvg_1_1HaplotypeIndexer.html#acb44a055288c0a9debed5662e461959e">More...</a><br /></td></tr>
<tr class="separator:acb44a055288c0a9debed5662e461959e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9991b56b5003075a5a072e06e669a3c1"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#a9991b56b5003075a5a072e06e669a3c1">discard_overlaps</a> = false</td></tr>
<tr class="memdesc:a9991b56b5003075a5a072e06e669a3c1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Join together overlapping haplotypes.  <a href="classvg_1_1HaplotypeIndexer.html#a9991b56b5003075a5a072e06e669a3c1">More...</a><br /></td></tr>
<tr class="separator:a9991b56b5003075a5a072e06e669a3c1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a681bcf3fb6014ae885adb3dec25ff2fc"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#a681bcf3fb6014ae885adb3dec25ff2fc">samples_in_batch</a> = 200</td></tr>
<tr class="memdesc:a681bcf3fb6014ae885adb3dec25ff2fc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of samples to process together in a haplotype batch.  <a href="classvg_1_1HaplotypeIndexer.html#a681bcf3fb6014ae885adb3dec25ff2fc">More...</a><br /></td></tr>
<tr class="separator:a681bcf3fb6014ae885adb3dec25ff2fc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abc8d5ce81546e16ffb7bf652c098bb11"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#abc8d5ce81546e16ffb7bf652c098bb11">gbwt_buffer_size</a> = gbwt::DynamicGBWT::INSERT_BATCH_SIZE / gbwt::MILLION</td></tr>
<tr class="memdesc:abc8d5ce81546e16ffb7bf652c098bb11"><td class="mdescLeft">&#160;</td><td class="mdescRight">Size of the GBWT buffer in millions of nodes.  <a href="classvg_1_1HaplotypeIndexer.html#abc8d5ce81546e16ffb7bf652c098bb11">More...</a><br /></td></tr>
<tr class="separator:abc8d5ce81546e16ffb7bf652c098bb11"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af71885cc768549e78097e76e77e90ea8"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#af71885cc768549e78097e76e77e90ea8">id_interval</a> = gbwt::DynamicGBWT::SAMPLE_INTERVAL</td></tr>
<tr class="memdesc:af71885cc768549e78097e76e77e90ea8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Interval at which to sample for GBWT locate.  <a href="classvg_1_1HaplotypeIndexer.html#af71885cc768549e78097e76e77e90ea8">More...</a><br /></td></tr>
<tr class="separator:af71885cc768549e78097e76e77e90ea8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a53cb273305c8fb1c039581c1a1b3d76e"><td class="memItemLeft" align="right" valign="top">std::pair&lt; size_t, size_t &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#a53cb273305c8fb1c039581c1a1b3d76e">sample_range</a> = std::pair&lt;size_t, size_t&gt;(0, std::numeric_limits&lt;size_t&gt;::max())</td></tr>
<tr class="memdesc:a53cb273305c8fb1c039581c1a1b3d76e"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structvg_1_1Range.html">Range</a> of VCF samples to process (first to past-last).  <a href="classvg_1_1HaplotypeIndexer.html#a53cb273305c8fb1c039581c1a1b3d76e">More...</a><br /></td></tr>
<tr class="separator:a53cb273305c8fb1c039581c1a1b3d76e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a85285e19121d708649f61b5420ab0777"><td class="memItemLeft" align="right" valign="top">std::map&lt; std::string, std::pair&lt; size_t, size_t &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#a85285e19121d708649f61b5420ab0777">regions</a></td></tr>
<tr class="separator:a85285e19121d708649f61b5420ab0777"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a116eb86acabeb52c7da7a1688686c034"><td class="memItemLeft" align="right" valign="top">std::unordered_set&lt; std::string &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1HaplotypeIndexer.html#a116eb86acabeb52c7da7a1688686c034">excluded_samples</a></td></tr>
<tr class="separator:a116eb86acabeb52c7da7a1688686c034"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="inherit_header pub_attribs_classvg_1_1Progressive"><td colspan="2" onclick="javascript:toggleInherit('pub_attribs_classvg_1_1Progressive')"><img src="closed.png" alt="-"/>&#160;Public Attributes inherited from <a class="el" href="classvg_1_1Progressive.html">vg::Progressive</a></td></tr>
<tr class="memitem:a6653625f496832c4cb8711185d5e04cb inherit pub_attribs_classvg_1_1Progressive"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Progressive.html#a6653625f496832c4cb8711185d5e04cb">show_progress</a> = false</td></tr>
<tr class="separator:a6653625f496832c4cb8711185d5e04cb inherit pub_attribs_classvg_1_1Progressive"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="inherited"></a>
Additional Inherited Members</h2></td></tr>
<tr class="inherit_header pub_static_methods_classvg_1_1Progressive"><td colspan="2" onclick="javascript:toggleInherit('pub_static_methods_classvg_1_1Progressive')"><img src="closed.png" alt="-"/>&#160;Static Public Member Functions inherited from <a class="el" href="classvg_1_1Progressive.html">vg::Progressive</a></td></tr>
<tr class="memitem:ab44ff020fc951b95dfcd1252c015a020 inherit pub_static_methods_classvg_1_1Progressive"><td class="memItemLeft" align="right" valign="top">static void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Progressive.html#ab44ff020fc951b95dfcd1252c015a020">with_progress</a> (bool <a class="el" href="classvg_1_1Progressive.html#a6653625f496832c4cb8711185d5e04cb">show_progress</a>, const std::string &amp;task, const std::function&lt; void(const std::function&lt; void(size_t, size_t)&gt; &amp;<a class="el" href="classvg_1_1Progressive.html#af78e64e28b252828a0ba86439125fea4">progress</a>)&gt; &amp;callback)</td></tr>
<tr class="separator:ab44ff020fc951b95dfcd1252c015a020 inherit pub_static_methods_classvg_1_1Progressive"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Allows indexing haplotypes, either to pre-parsed haplotype files or to a GBWT. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="ae0b8cbfe64f32152ceefa57daf9ece72"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae0b8cbfe64f32152ceefa57daf9ece72">&#9670;&nbsp;</a></span>HaplotypeIndexer()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vg::HaplotypeIndexer::HaplotypeIndexer </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Perform initialization of backing libraries. </p>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="a34fc36c5ba77f4cedd32073dda118b84"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a34fc36c5ba77f4cedd32073dda118b84">&#9670;&nbsp;</a></span>build_gbwt() <span class="overload">[1/3]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::unique_ptr&lt; gbwt::GBWT &gt; vg::HaplotypeIndexer::build_gbwt </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classhandlegraph_1_1HandleGraph.html">HandleGraph</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::vector&lt; std::string &gt; &amp;&#160;</td>
          <td class="paramname"><em>aln_filenames</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>aln_format</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>parallel_jobs</em> = <code>1</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Build a GBWT from the alignments. Each distinct alignment name becomes a sample in the GBWT metadata. If there are multiple alignments with the same name, the corresponding GBWT path names will have the same sample identifier but different values in the count field.</p>
<p>aln_format can be "GAM" or "GAF".</p>
<p>Runs approximately the given number of jobs in parallel. The exact number depends on the sizes of weakly connected components in the graph. Each job uses at most 2 threads. </p>

</div>
</div>
<a id="ae92d109c2e7baf0b7ddded8df80a6f70"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae92d109c2e7baf0b7ddded8df80a6f70">&#9670;&nbsp;</a></span>build_gbwt() <span class="overload">[2/3]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::unique_ptr&lt; gbwt::DynamicGBWT &gt; vg::HaplotypeIndexer::build_gbwt </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classhandlegraph_1_1PathHandleGraph.html">PathHandleGraph</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Build a GBWT from the embedded non-alt paths in the graph. </p>

</div>
</div>
<a id="a0389ee96c5ce8a6d9c158fda39aca9d9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0389ee96c5ce8a6d9c158fda39aca9d9">&#9670;&nbsp;</a></span>build_gbwt() <span class="overload">[3/3]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::unique_ptr&lt; gbwt::DynamicGBWT &gt; vg::HaplotypeIndexer::build_gbwt </td>
          <td>(</td>
          <td class="paramtype">const std::vector&lt; std::string &gt; &amp;&#160;</td>
          <td class="paramname"><em>vcf_parse_files</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>job_name</em> = <code>&quot;GBWT&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classhandlegraph_1_1PathHandleGraph.html">PathHandleGraph</a> *&#160;</td>
          <td class="paramname"><em>graph</em> = <code>nullptr</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::unordered_set&lt; <a class="el" href="structhandlegraph_1_1path__handle__t.html">path_handle_t</a> &gt; *&#160;</td>
          <td class="paramname"><em>paths</em> = <code>nullptr</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>skip_unvisited_paths</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Build a GBWT from the haplotypes in the given VCF parse files.</p>
<p>Respects excluded_samples and does not produce threads for them.</p>
<p>We expect that all parse files contain sample/contig names and that the sample names are the same in all files.</p>
<p>There may be no parse files.</p>
<p>If graph is provided and is not null, also includes embedded non-alt paths from the graph.</p>
<p>If paths is provided and is not null, include only those specified paths from the graph. If skip_unvisited_paths is set, paths whose contigs are not visited by VCF parse files will be skipped. </p>

</div>
</div>
<a id="ad3eda4acd04b8bae303b4604410ece41"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad3eda4acd04b8bae303b4604410ece41">&#9670;&nbsp;</a></span>parse_vcf() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt; std::string &gt; vg::HaplotypeIndexer::parse_vcf </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classhandlegraph_1_1PathHandleGraph.html">PathHandleGraph</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>job_name</em> = <code>&quot;GBWT&quot;</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Parse the VCF file into the types needed for GBWT indexing.</p>
<p>Returns the file names for the VCF parses of non-alt paths. If batch_file_prefix is set, these are permanent files. Otherwise they are temporary files that persist until the program exits. </p>

</div>
</div>
<a id="ad81dbd24bba40aaa14536d79754d387b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad81dbd24bba40aaa14536d79754d387b">&#9670;&nbsp;</a></span>parse_vcf() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt; std::string &gt; vg::HaplotypeIndexer::parse_vcf </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classhandlegraph_1_1PathHandleGraph.html">PathHandleGraph</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::vector&lt; <a class="el" href="structhandlegraph_1_1path__handle__t.html">path_handle_t</a> &gt; &amp;&#160;</td>
          <td class="paramname"><em>paths</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>job_name</em> = <code>&quot;GBWT&quot;</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Parse the VCF file into the types needed for GBWT indexing.</p>
<p>Returns the file names for the VCF parses of the specified paths. If batch_file_prefix is set, these are permanent files. Otherwise they are temporary files that persist until the program exits. </p>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a id="aa818eebd7cffc81773e8bea65bd398cb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa818eebd7cffc81773e8bea65bd398cb">&#9670;&nbsp;</a></span>batch_file_prefix</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::string vg::HaplotypeIndexer::batch_file_prefix = &quot;&quot;</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>If batch_file_prefix is nonempty, a file for each contig is saved to PREFIX_VCFCONTIG, and files for each batch of haplotypes are saved to files named like PREFIX_VCFCONTIG_STARTSAMPLE_ENDSAMPLE. Otherwise, the batch files are still saved, but to temporary files. </p>

</div>
</div>
<a id="a9991b56b5003075a5a072e06e669a3c1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9991b56b5003075a5a072e06e669a3c1">&#9670;&nbsp;</a></span>discard_overlaps</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::HaplotypeIndexer::discard_overlaps = false</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Join together overlapping haplotypes. </p>

</div>
</div>
<a id="a116eb86acabeb52c7da7a1688686c034"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a116eb86acabeb52c7da7a1688686c034">&#9670;&nbsp;</a></span>excluded_samples</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::unordered_set&lt;std::string&gt; vg::HaplotypeIndexer::excluded_samples</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Excluded VCF sample names, for which threads will not be generated. Ignored during VCF parsing. </p>

</div>
</div>
<a id="acb44a055288c0a9debed5662e461959e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acb44a055288c0a9debed5662e461959e">&#9670;&nbsp;</a></span>force_phasing</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::HaplotypeIndexer::force_phasing = false</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Arbitrarily phase all unphased variants. </p>

</div>
</div>
<a id="abc8d5ce81546e16ffb7bf652c098bb11"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abc8d5ce81546e16ffb7bf652c098bb11">&#9670;&nbsp;</a></span>gbwt_buffer_size</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t vg::HaplotypeIndexer::gbwt_buffer_size = gbwt::DynamicGBWT::INSERT_BATCH_SIZE / gbwt::MILLION</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Size of the GBWT buffer in millions of nodes. </p>

</div>
</div>
<a id="af71885cc768549e78097e76e77e90ea8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af71885cc768549e78097e76e77e90ea8">&#9670;&nbsp;</a></span>id_interval</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t vg::HaplotypeIndexer::id_interval = gbwt::DynamicGBWT::SAMPLE_INTERVAL</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Interval at which to sample for GBWT locate. </p>

</div>
</div>
<a id="a512203218e3b78afa0fc5c0e594e6516"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a512203218e3b78afa0fc5c0e594e6516">&#9670;&nbsp;</a></span>max_missing_variant_warnings</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t vg::HaplotypeIndexer::max_missing_variant_warnings = 10</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Only report up to this many of them. </p>

</div>
</div>
<a id="a5ff016a7fdf425ec4f3d5408d09b951f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5ff016a7fdf425ec4f3d5408d09b951f">&#9670;&nbsp;</a></span>path_to_vcf</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::map&lt;std::string, std::string&gt; vg::HaplotypeIndexer::path_to_vcf</td>
        </tr>
      </table>
</div><div class="memdoc">
<p><a class="el" href="structvg_1_1Path.html">Path</a> names in the graph are mapped to VCF contig names via path_to_vcf, or used as-is if no entry there is found. </p>

</div>
</div>
<a id="a881e755025092748b18f037fb1078e65"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a881e755025092748b18f037fb1078e65">&#9670;&nbsp;</a></span>phase_homozygous</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::HaplotypeIndexer::phase_homozygous = true</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Phase homozygous unphased variants. </p>

</div>
</div>
<a id="a85285e19121d708649f61b5420ab0777"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a85285e19121d708649f61b5420ab0777">&#9670;&nbsp;</a></span>regions</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::map&lt;std::string, std::pair&lt;size_t, size_t&gt; &gt; vg::HaplotypeIndexer::regions</td>
        </tr>
      </table>
</div><div class="memdoc">
<p><a class="el" href="structvg_1_1Region.html">Region</a> restrictions for contigs, in VCF name space, as 0-based exclusive-end ranges. </p>

</div>
</div>
<a id="a8061130697dc26811303c4b67b081245"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8061130697dc26811303c4b67b081245">&#9670;&nbsp;</a></span>rename_variants</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::HaplotypeIndexer::rename_variants = true</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Use graph path names instead of VCF path names when composing variant alt paths. </p>

</div>
</div>
<a id="a53cb273305c8fb1c039581c1a1b3d76e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a53cb273305c8fb1c039581c1a1b3d76e">&#9670;&nbsp;</a></span>sample_range</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::pair&lt;size_t, size_t&gt; vg::HaplotypeIndexer::sample_range = std::pair&lt;size_t, size_t&gt;(0, std::numeric_limits&lt;size_t&gt;::max())</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><a class="el" href="structvg_1_1Range.html">Range</a> of VCF samples to process (first to past-last). </p>

</div>
</div>
<a id="a681bcf3fb6014ae885adb3dec25ff2fc"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a681bcf3fb6014ae885adb3dec25ff2fc">&#9670;&nbsp;</a></span>samples_in_batch</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t vg::HaplotypeIndexer::samples_in_batch = 200</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Number of samples to process together in a haplotype batch. </p>

</div>
</div>
<a id="afb3592792a7d781cb6e88fe7949b7bfe"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afb3592792a7d781cb6e88fe7949b7bfe">&#9670;&nbsp;</a></span>warn_on_missing_variants</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::HaplotypeIndexer::warn_on_missing_variants = true</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Print a warning if variants in the VCF can't be found in the graph. </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>src/<a class="el" href="haplotype__indexer_8hpp.html">haplotype_indexer.hpp</a></li>
<li>src/<a class="el" href="haplotype__indexer_8cpp.html">haplotype_indexer.cpp</a></li>
</ul>
</div><!-- contents -->
<!-- HTML footer for doxygen 1.8.9.1-->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.17
</small></address>
</body>
</html>
