<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.17"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>vg: vg::Haplotypes Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">vg
   </div>
   <div id="projectbrief">tools for working with variation graphs</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.17 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacevg.html">vg</a></li><li class="navelem"><a class="el" href="classvg_1_1Haplotypes.html">Haplotypes</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-types">Public Types</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="classvg_1_1Haplotypes-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">vg::Haplotypes Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p><code>#include &lt;recombinator.hpp&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Haplotypes_1_1Header.html">Header</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structvg_1_1Haplotypes_1_1Header.html" title="Header of the serialized file.">Header</a> of the serialized file.  <a href="structvg_1_1Haplotypes_1_1Header.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Haplotypes_1_1Subchain.html">Subchain</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Representation of a subchain.  <a href="structvg_1_1Haplotypes_1_1Subchain.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Haplotypes_1_1TopLevelChain.html">TopLevelChain</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Representation of a top-level chain.  <a href="structvg_1_1Haplotypes_1_1TopLevelChain.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-types"></a>
Public Types</h2></td></tr>
<tr class="memitem:a2e2fbea03cc9fb69389a653d3d8fc28a"><td class="memItemLeft" align="right" valign="top">enum &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#a2e2fbea03cc9fb69389a653d3d8fc28a">Verbosity</a> : size_t { <br />
&#160;&#160;<a class="el" href="classvg_1_1Haplotypes.html#a2e2fbea03cc9fb69389a653d3d8fc28aadd4472c60bf6446f52a09a0d41140a21">verbosity_silent</a> = 0, 
<a class="el" href="classvg_1_1Haplotypes.html#a2e2fbea03cc9fb69389a653d3d8fc28aa718364c23222ec5e7444240be0799449">verbosity_basic</a> = 1, 
<a class="el" href="classvg_1_1Haplotypes.html#a2e2fbea03cc9fb69389a653d3d8fc28aa8c3562efded9b2b9ae9fd7ee05594ce0">verbosity_detailed</a> = 2, 
<a class="el" href="classvg_1_1Haplotypes.html#a2e2fbea03cc9fb69389a653d3d8fc28aaa92669d64637a1583a113eb279234241">verbosity_debug</a> = 3, 
<br />
&#160;&#160;<a class="el" href="classvg_1_1Haplotypes.html#a2e2fbea03cc9fb69389a653d3d8fc28aa08dcddd51ddf6c9e8df59757e13eaa30">verbosity_extra_debug</a> = 4
<br />
 }</td></tr>
<tr class="memdesc:a2e2fbea03cc9fb69389a653d3d8fc28a"><td class="mdescLeft">&#160;</td><td class="mdescRight">The amount of progress information that should be printed to stderr.  <a href="classvg_1_1Haplotypes.html#a2e2fbea03cc9fb69389a653d3d8fc28a">More...</a><br /></td></tr>
<tr class="separator:a2e2fbea03cc9fb69389a653d3d8fc28a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aebc70f387d48bd0e5c8f0c607c496535"><td class="memItemLeft" align="right" valign="top">typedef std::pair&lt; gbwt::size_type, gbwt::size_type &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#aebc70f387d48bd0e5c8f0c607c496535">sequence_type</a></td></tr>
<tr class="memdesc:aebc70f387d48bd0e5c8f0c607c496535"><td class="mdescLeft">&#160;</td><td class="mdescRight">A GBWT sequence as (sequence identifier, offset in a node).  <a href="classvg_1_1Haplotypes.html#aebc70f387d48bd0e5c8f0c607c496535">More...</a><br /></td></tr>
<tr class="separator:aebc70f387d48bd0e5c8f0c607c496535"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa032022df55d45eed0ad1c509a95a5c7"><td class="memItemLeft" align="right" valign="top">typedef std::pair&lt; std::uint32_t, std::uint32_t &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#aa032022df55d45eed0ad1c509a95a5c7">compact_sequence_type</a></td></tr>
<tr class="memdesc:aa032022df55d45eed0ad1c509a95a5c7"><td class="mdescLeft">&#160;</td><td class="mdescRight">A more space-efficient representation of <code>sequence_type</code>.  <a href="classvg_1_1Haplotypes.html#aa032022df55d45eed0ad1c509a95a5c7">More...</a><br /></td></tr>
<tr class="separator:aa032022df55d45eed0ad1c509a95a5c7"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a1617e9b0fbd01fc77fd87c527e5dbd24"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#a1617e9b0fbd01fc77fd87c527e5dbd24">components</a> () const</td></tr>
<tr class="memdesc:a1617e9b0fbd01fc77fd87c527e5dbd24"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the number of weakly connected components.  <a href="classvg_1_1Haplotypes.html#a1617e9b0fbd01fc77fd87c527e5dbd24">More...</a><br /></td></tr>
<tr class="separator:a1617e9b0fbd01fc77fd87c527e5dbd24"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af1b40bc51ea14870df84ab86ca71c0ce"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#af1b40bc51ea14870df84ab86ca71c0ce">jobs</a> () const</td></tr>
<tr class="memdesc:af1b40bc51ea14870df84ab86ca71c0ce"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the number of GBWT construction jobs.  <a href="classvg_1_1Haplotypes.html#af1b40bc51ea14870df84ab86ca71c0ce">More...</a><br /></td></tr>
<tr class="separator:af1b40bc51ea14870df84ab86ca71c0ce"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af081161e7b2e1867d83aef6427082cd5"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#af081161e7b2e1867d83aef6427082cd5">k</a> () const</td></tr>
<tr class="memdesc:af081161e7b2e1867d83aef6427082cd5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the length of the kmers.  <a href="classvg_1_1Haplotypes.html#af081161e7b2e1867d83aef6427082cd5">More...</a><br /></td></tr>
<tr class="separator:af081161e7b2e1867d83aef6427082cd5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a85b9a6d31a3485f0ccf925023543414b"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#a85b9a6d31a3485f0ccf925023543414b">kmers</a> () const</td></tr>
<tr class="memdesc:a85b9a6d31a3485f0ccf925023543414b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the number of kmers in the subchains.  <a href="classvg_1_1Haplotypes.html#a85b9a6d31a3485f0ccf925023543414b">More...</a><br /></td></tr>
<tr class="separator:a85b9a6d31a3485f0ccf925023543414b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a48ed72eff81d77df7a4159afd0c6d20e"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classvg_1_1hash__map.html">hash_map</a>&lt; <a class="el" href="structvg_1_1Haplotypes_1_1Subchain.html#a13b4b76384742f4ee127b52faff37aa6">Subchain::kmer_type</a>, size_t &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#a48ed72eff81d77df7a4159afd0c6d20e">kmer_counts</a> (const std::string &amp;kff_file, <a class="el" href="classvg_1_1Haplotypes.html#a2e2fbea03cc9fb69389a653d3d8fc28a">Verbosity</a> verbosity) const</td></tr>
<tr class="separator:a48ed72eff81d77df7a4159afd0c6d20e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3a5a36c107109c893cf382064ad9ea69"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#a3a5a36c107109c893cf382064ad9ea69">simple_sds_serialize</a> (std::ostream &amp;out) const</td></tr>
<tr class="separator:a3a5a36c107109c893cf382064ad9ea69"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a59cabeadc7a097f9c1926bff45a77505"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#a59cabeadc7a097f9c1926bff45a77505">serialize_to</a> (const std::string &amp;filename) const</td></tr>
<tr class="separator:a59cabeadc7a097f9c1926bff45a77505"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a908ed1a2f1147a2162051be87e5ba389"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#a908ed1a2f1147a2162051be87e5ba389">simple_sds_load</a> (std::istream &amp;in)</td></tr>
<tr class="separator:a908ed1a2f1147a2162051be87e5ba389"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae4ce505e821cdd049f26dc932d6bfc2c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#ae4ce505e821cdd049f26dc932d6bfc2c">load_from</a> (const std::string &amp;filename)</td></tr>
<tr class="separator:ae4ce505e821cdd049f26dc932d6bfc2c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac0ebaa605235ee2b6f62c084cfdc6126"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#ac0ebaa605235ee2b6f62c084cfdc6126">simple_sds_size</a> () const</td></tr>
<tr class="memdesc:ac0ebaa605235ee2b6f62c084cfdc6126"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the size of the object in elements.  <a href="classvg_1_1Haplotypes.html#ac0ebaa605235ee2b6f62c084cfdc6126">More...</a><br /></td></tr>
<tr class="separator:ac0ebaa605235ee2b6f62c084cfdc6126"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a68d4b34e9075418b258faefd221c71ed"><td class="memItemLeft" align="right" valign="top">std::vector&lt; size_t &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#a68d4b34e9075418b258faefd221c71ed">assign_reference_paths</a> (const gbwtgraph::GBZ &amp;gbz, <a class="el" href="classvg_1_1Haplotypes.html#a2e2fbea03cc9fb69389a653d3d8fc28a">Verbosity</a> verbosity) const</td></tr>
<tr class="separator:a68d4b34e9075418b258faefd221c71ed"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:ac2671cb0e5d0cdf19a8aa849ec9a2577"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structvg_1_1Haplotypes_1_1Header.html">Header</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#ac2671cb0e5d0cdf19a8aa849ec9a2577">header</a></td></tr>
<tr class="separator:ac2671cb0e5d0cdf19a8aa849ec9a2577"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9f3d8db67fd1488075960f77b361c161"><td class="memItemLeft" align="right" valign="top">std::vector&lt; size_t &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#a9f3d8db67fd1488075960f77b361c161">jobs_for_paths</a></td></tr>
<tr class="separator:a9f3d8db67fd1488075960f77b361c161"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aabf788ef8783a229fa095209bf61d7eb"><td class="memItemLeft" align="right" valign="top">std::vector&lt; <a class="el" href="structvg_1_1Haplotypes_1_1TopLevelChain.html">TopLevelChain</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Haplotypes.html#aabf788ef8783a229fa095209bf61d7eb">chains</a></td></tr>
<tr class="separator:aabf788ef8783a229fa095209bf61d7eb"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>A representation of the haplotypes in a graph.</p>
<p>The graph is partitioned into top-level chains, which are further partitioned into subchains. Each subchain contains a set of kmers and a collection of sequences. Each sequence is defined by a bitvector marking the kmers that are present.</p>
<p>At the moment, the kmers are minimizers with a single occurrence in the graph. The requirement is that each kmer is specific to a single subchain and does not occur anywhere else in either orientation. (If no haplotype crosses a snarl, that snarl is broken into a suffix and a prefix, and those subchains may share kmers.)</p>
<p>NOTE: This assumes that the top-level chains are linear, not cyclical.</p>
<p>Versions:</p>
<ul>
<li><a class="el" href="classvg_1_1Version.html" title="Class for holding vg version and build environment information.">Version</a> 5: Every path in the graph is assigned to a construction job. This allows including reference paths that do not visit any snarls in the sampled graph. Not compatible with earlier versions.</li>
<li><a class="el" href="classvg_1_1Version.html" title="Class for holding vg version and build environment information.">Version</a> 4: Subchains can have fragmented haplotypes instead of a single GBWT sequence always crossing from start to end. Compatible with version 3.</li>
<li><a class="el" href="classvg_1_1Version.html" title="Class for holding vg version and build environment information.">Version</a> 3: Subchains use smaller integers when possible. Compatible with version 2.</li>
<li><a class="el" href="classvg_1_1Version.html" title="Class for holding vg version and build environment information.">Version</a> 2: Top-level chains include a contig name. Compatible with version 1.</li>
<li><a class="el" href="classvg_1_1Version.html" title="Class for holding vg version and build environment information.">Version</a> 1: Initial version. </li>
</ul>
</div><h2 class="groupheader">Member Typedef Documentation</h2>
<a id="aa032022df55d45eed0ad1c509a95a5c7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa032022df55d45eed0ad1c509a95a5c7">&#9670;&nbsp;</a></span>compact_sequence_type</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">typedef std::pair&lt;std::uint32_t, std::uint32_t&gt; <a class="el" href="classvg_1_1Haplotypes.html#aa032022df55d45eed0ad1c509a95a5c7">vg::Haplotypes::compact_sequence_type</a></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>A more space-efficient representation of <code>sequence_type</code>. </p>

</div>
</div>
<a id="aebc70f387d48bd0e5c8f0c607c496535"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aebc70f387d48bd0e5c8f0c607c496535">&#9670;&nbsp;</a></span>sequence_type</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">typedef std::pair&lt;gbwt::size_type, gbwt::size_type&gt; <a class="el" href="classvg_1_1Haplotypes.html#aebc70f387d48bd0e5c8f0c607c496535">vg::Haplotypes::sequence_type</a></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>A GBWT sequence as (sequence identifier, offset in a node). </p>

</div>
</div>
<h2 class="groupheader">Member Enumeration Documentation</h2>
<a id="a2e2fbea03cc9fb69389a653d3d8fc28a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2e2fbea03cc9fb69389a653d3d8fc28a">&#9670;&nbsp;</a></span>Verbosity</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">enum <a class="el" href="classvg_1_1Haplotypes.html#a2e2fbea03cc9fb69389a653d3d8fc28a">vg::Haplotypes::Verbosity</a> : size_t</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The amount of progress information that should be printed to stderr. </p>
<table class="fieldtable">
<tr><th colspan="2">Enumerator</th></tr><tr><td class="fieldname"><a id="a2e2fbea03cc9fb69389a653d3d8fc28aadd4472c60bf6446f52a09a0d41140a21"></a>verbosity_silent&#160;</td><td class="fielddoc"><p>No progress information. </p>
</td></tr>
<tr><td class="fieldname"><a id="a2e2fbea03cc9fb69389a653d3d8fc28aa718364c23222ec5e7444240be0799449"></a>verbosity_basic&#160;</td><td class="fielddoc"><p>Basic information. </p>
</td></tr>
<tr><td class="fieldname"><a id="a2e2fbea03cc9fb69389a653d3d8fc28aa8c3562efded9b2b9ae9fd7ee05594ce0"></a>verbosity_detailed&#160;</td><td class="fielddoc"><p>Basic information and detailed statistics. </p>
</td></tr>
<tr><td class="fieldname"><a id="a2e2fbea03cc9fb69389a653d3d8fc28aaa92669d64637a1583a113eb279234241"></a>verbosity_debug&#160;</td><td class="fielddoc"><p>Basic information, detailed statistics, and debug information. </p>
</td></tr>
<tr><td class="fieldname"><a id="a2e2fbea03cc9fb69389a653d3d8fc28aa08dcddd51ddf6c9e8df59757e13eaa30"></a>verbosity_extra_debug&#160;</td><td class="fielddoc"><p>Hidden level; potentially tens of thousands of lines of debugging information. </p>
</td></tr>
</table>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="a68d4b34e9075418b258faefd221c71ed"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a68d4b34e9075418b258faefd221c71ed">&#9670;&nbsp;</a></span>assign_reference_paths()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt; size_t &gt; vg::Haplotypes::assign_reference_paths </td>
          <td>(</td>
          <td class="paramtype">const gbwtgraph::GBZ &amp;&#160;</td>
          <td class="paramname"><em>gbz</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1Haplotypes.html#a2e2fbea03cc9fb69389a653d3d8fc28a">Verbosity</a>&#160;</td>
          <td class="paramname"><em>verbosity</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Assigns each reference and generic path in the graph to a GBWT construction job.</p>
<p>For each path handle from 0 to gbz.named_paths() - 1, we assign the path to the given construction job, or <a class="el" href="classvg_1_1Haplotypes.html#af1b40bc51ea14870df84ab86ca71c0ce" title="Returns the number of GBWT construction jobs.">jobs()</a> if the path is empty. </p>

</div>
</div>
<a id="a1617e9b0fbd01fc77fd87c527e5dbd24"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1617e9b0fbd01fc77fd87c527e5dbd24">&#9670;&nbsp;</a></span>components()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">size_t vg::Haplotypes::components </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Returns the number of weakly connected components. </p>

</div>
</div>
<a id="af1b40bc51ea14870df84ab86ca71c0ce"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af1b40bc51ea14870df84ab86ca71c0ce">&#9670;&nbsp;</a></span>jobs()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">size_t vg::Haplotypes::jobs </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Returns the number of GBWT construction jobs. </p>

</div>
</div>
<a id="af081161e7b2e1867d83aef6427082cd5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af081161e7b2e1867d83aef6427082cd5">&#9670;&nbsp;</a></span>k()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">size_t vg::Haplotypes::k </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Returns the length of the kmers. </p>

</div>
</div>
<a id="a48ed72eff81d77df7a4159afd0c6d20e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a48ed72eff81d77df7a4159afd0c6d20e">&#9670;&nbsp;</a></span>kmer_counts()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classvg_1_1hash__map.html">hash_map</a>&lt; <a class="el" href="structvg_1_1Haplotypes_1_1Subchain.html#a13b4b76384742f4ee127b52faff37aa6">Haplotypes::Subchain::kmer_type</a>, size_t &gt; vg::Haplotypes::kmer_counts </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>kff_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1Haplotypes.html#a2e2fbea03cc9fb69389a653d3d8fc28a">Verbosity</a>&#160;</td>
          <td class="paramname"><em>verbosity</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Returns a mapping from kmers to their counts in the given KFF file. The counts include both the kmer and the reverse complement.</p>
<p>Reads the KFF file using OpenMP threads. Exits with <code>std::exit()</code> if the file cannot be opened and throws <code>std::runtime_error</code> if the kmer counts cannot be used. </p>

</div>
</div>
<a id="a85b9a6d31a3485f0ccf925023543414b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a85b9a6d31a3485f0ccf925023543414b">&#9670;&nbsp;</a></span>kmers()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">size_t vg::Haplotypes::kmers </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Returns the number of kmers in the subchains. </p>

</div>
</div>
<a id="ae4ce505e821cdd049f26dc932d6bfc2c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae4ce505e821cdd049f26dc932d6bfc2c">&#9670;&nbsp;</a></span>load_from()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Haplotypes::load_from </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>filename</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Loads the object from a file in the Simple-SDS format. Prints an error message and exits the program on failure. </p>

</div>
</div>
<a id="a59cabeadc7a097f9c1926bff45a77505"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a59cabeadc7a097f9c1926bff45a77505">&#9670;&nbsp;</a></span>serialize_to()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Haplotypes::serialize_to </td>
          <td>(</td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>filename</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Serializes the object to a file in the Simple-SDS format. Prints an error message and exits the program on failure. </p>

</div>
</div>
<a id="a908ed1a2f1147a2162051be87e5ba389"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a908ed1a2f1147a2162051be87e5ba389">&#9670;&nbsp;</a></span>simple_sds_load()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Haplotypes::simple_sds_load </td>
          <td>(</td>
          <td class="paramtype">std::istream &amp;&#160;</td>
          <td class="paramname"><em>in</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Loads the object from a stream in the Simple-SDS format. I/O errors can be detected by checking the stream state. Throws <code>sdsl::simple_sds::InvalidData</code> if data is unacceptable. </p>

</div>
</div>
<a id="a3a5a36c107109c893cf382064ad9ea69"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3a5a36c107109c893cf382064ad9ea69">&#9670;&nbsp;</a></span>simple_sds_serialize()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Haplotypes::simple_sds_serialize </td>
          <td>(</td>
          <td class="paramtype">std::ostream &amp;&#160;</td>
          <td class="paramname"><em>out</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Serializes the object to a stream in the Simple-SDS format. I/O errors can be detected by checking the stream state. </p>

</div>
</div>
<a id="ac0ebaa605235ee2b6f62c084cfdc6126"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac0ebaa605235ee2b6f62c084cfdc6126">&#9670;&nbsp;</a></span>simple_sds_size()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t vg::Haplotypes::simple_sds_size </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the size of the object in elements. </p>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a id="aabf788ef8783a229fa095209bf61d7eb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aabf788ef8783a229fa095209bf61d7eb">&#9670;&nbsp;</a></span>chains</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;<a class="el" href="structvg_1_1Haplotypes_1_1TopLevelChain.html">TopLevelChain</a>&gt; vg::Haplotypes::chains</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="ac2671cb0e5d0cdf19a8aa849ec9a2577"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac2671cb0e5d0cdf19a8aa849ec9a2577">&#9670;&nbsp;</a></span>header</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structvg_1_1Haplotypes_1_1Header.html">Header</a> vg::Haplotypes::header</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a9f3d8db67fd1488075960f77b361c161"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9f3d8db67fd1488075960f77b361c161">&#9670;&nbsp;</a></span>jobs_for_paths</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt;size_t&gt; vg::Haplotypes::jobs_for_paths</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>src/<a class="el" href="recombinator_8hpp.html">recombinator.hpp</a></li>
<li>src/<a class="el" href="recombinator_8cpp.html">recombinator.cpp</a></li>
</ul>
</div><!-- contents -->
<!-- HTML footer for doxygen 1.8.9.1-->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.17
</small></address>
</body>
</html>
