<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>vg: vg::Genotyper Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">vg
   </div>
   <div id="projectbrief">tools for working with variation graphs</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacevg.html">vg</a></li><li class="navelem"><a class="el" href="classvg_1_1Genotyper.html">Genotyper</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-types">Public Types</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="classvg_1_1Genotyper-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">vg::Genotyper Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p><code>#include &lt;genotyper.hpp&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Genotyper_1_1Affinity.html">Affinity</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Genotyper_1_1hash__ambiguous__allele__set.html">hash_ambiguous_allele_set</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Genotyper_1_1hash__node__traversal.html">hash_node_traversal</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Genotyper_1_1hash__oriented__edge.html">hash_oriented_edge</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-types"></a>
Public Types</h2></td></tr>
<tr class="memitem:a6e4dce6cd9cc24ab6857ceeaed6db3b4"><td class="memItemLeft" align="right" valign="top">enum &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4">TraversalAlg</a> { <a class="el" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4a9989ae14c6b6c27998126ab745488926">Reads</a>
, <a class="el" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4a03b965f4f4d03d6869811e0dedbfb0dd">Exhaustive</a>
, <a class="el" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4aca000fa2f53ba37764dedd204e1fd27e">Representative</a>
, <a class="el" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4ab530b2114d4cd7f36bb3a41c01928995">Adaptive</a>
 }</td></tr>
<tr class="separator:a6e4dce6cd9cc24ab6857ceeaed6db3b4"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:af760be5c99e086b2c41510b3d3a47e5a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#af760be5c99e086b2c41510b3d3a47e5a">run</a> (<a class="el" href="structvg_1_1AugmentedGraph.html">AugmentedGraph</a> &amp;graph, ostream &amp;out, string ref_path_name, string contig_name=&quot;&quot;, string sample_name=&quot;&quot;, string augmented_file_name=&quot;&quot;, bool output_vcf=false, bool output_json=false, int length_override=0, int variant_offset=0)</td></tr>
<tr class="separator:af760be5c99e086b2c41510b3d3a47e5a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a76bba1da02c9a75895896efd5b5e3641"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a76bba1da02c9a75895896efd5b5e3641">alignment_qual_score</a> (<a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph, const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *snarl, const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> &amp;alignment)</td></tr>
<tr class="separator:a76bba1da02c9a75895896efd5b5e3641"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7a14f1e259fe612b944aadf701ca1585"><td class="memItemLeft" align="right" valign="top">vector&lt; <a class="el" href="structvg_1_1SnarlTraversal.html">SnarlTraversal</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a7a14f1e259fe612b944aadf701ca1585">get_snarl_traversals</a> (<a class="el" href="structvg_1_1AugmentedGraph.html">AugmentedGraph</a> &amp;augmented_graph, <a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a> &amp;manager, map&lt; string, const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> * &gt; &amp;reads_by_name, const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *snarl, const pair&lt; unordered_set&lt; <a class="el" href="namespacevg.html#afaa15fd8bdfd688d48561e1b12aee821">id_t</a> &gt;, unordered_set&lt; <a class="el" href="namespacevg.html#a3de8fe5f38b1730a36af4ad9c4258963">edge_t</a> &gt; &gt; &amp;contents, <a class="el" href="structvg_1_1PathIndex.html">PathIndex</a> *reference_index, <a class="el" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4">TraversalAlg</a> use_traversal_alg)</td></tr>
<tr class="separator:a7a14f1e259fe612b944aadf701ca1585"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af6518141a6e6f122e5f534c041d16918"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#af6518141a6e6f122e5f534c041d16918">get_qualities_in_snarl</a> (<a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph, const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *snarl, const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> &amp;alignment)</td></tr>
<tr class="separator:af6518141a6e6f122e5f534c041d16918"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abea795c1969c4866c349a07c7f9c7102"><td class="memItemLeft" align="right" valign="top">map&lt; const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> *, vector&lt; <a class="el" href="structvg_1_1Genotyper_1_1Affinity.html">Affinity</a> &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#abea795c1969c4866c349a07c7f9c7102">get_affinities</a> (<a class="el" href="structvg_1_1AugmentedGraph.html">AugmentedGraph</a> &amp;aug, const map&lt; string, const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> * &gt; &amp;reads_by_name, const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *snarl, const pair&lt; unordered_set&lt; <a class="el" href="namespacevg.html#afaa15fd8bdfd688d48561e1b12aee821">id_t</a> &gt;, unordered_set&lt; <a class="el" href="namespacevg.html#a3de8fe5f38b1730a36af4ad9c4258963">edge_t</a> &gt; &gt; &amp;contents, const <a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a> &amp;manager, const vector&lt; <a class="el" href="structvg_1_1SnarlTraversal.html">SnarlTraversal</a> &gt; &amp;superbubble_paths)</td></tr>
<tr class="separator:abea795c1969c4866c349a07c7f9c7102"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a966eee3f764281271904996546ef205b"><td class="memItemLeft" align="right" valign="top">map&lt; const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> *, vector&lt; <a class="el" href="structvg_1_1Genotyper_1_1Affinity.html">Affinity</a> &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a966eee3f764281271904996546ef205b">get_affinities_fast</a> (<a class="el" href="structvg_1_1AugmentedGraph.html">AugmentedGraph</a> &amp;aug, const map&lt; string, const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> * &gt; &amp;reads_by_name, const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *snarl, const pair&lt; unordered_set&lt; <a class="el" href="namespacevg.html#afaa15fd8bdfd688d48561e1b12aee821">id_t</a> &gt;, unordered_set&lt; <a class="el" href="namespacevg.html#a3de8fe5f38b1730a36af4ad9c4258963">edge_t</a> &gt; &gt; &amp;contents, const <a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a> &amp;manager, const vector&lt; <a class="el" href="structvg_1_1SnarlTraversal.html">SnarlTraversal</a> &gt; &amp;superbubble_paths, bool allow_internal_alignments=false)</td></tr>
<tr class="separator:a966eee3f764281271904996546ef205b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae6a3a060adcffc32724498bbfda2c771"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structvg_1_1Locus.html">Locus</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#ae6a3a060adcffc32724498bbfda2c771">genotype_snarl</a> (<a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph, const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *snarl, const vector&lt; <a class="el" href="structvg_1_1SnarlTraversal.html">SnarlTraversal</a> &gt; &amp;superbubble_paths, const map&lt; const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> *, vector&lt; <a class="el" href="structvg_1_1Genotyper_1_1Affinity.html">Affinity</a> &gt;&gt; &amp;affinities)</td></tr>
<tr class="separator:ae6a3a060adcffc32724498bbfda2c771"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9071cdf5ae591b478feaac025b069772"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a9071cdf5ae591b478feaac025b069772">get_genotype_log_likelihood</a> (<a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph, const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *snarl, const vector&lt; int &gt; &amp;genotype, const vector&lt; pair&lt; const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> *, vector&lt; <a class="el" href="structvg_1_1Genotyper_1_1Affinity.html">Affinity</a> &gt;&gt;&gt; &amp;alignment_consistency)</td></tr>
<tr class="separator:a9071cdf5ae591b478feaac025b069772"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5fc66cb05c4ba80e39d078af9ddce715"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a5fc66cb05c4ba80e39d078af9ddce715">get_genotype_log_prior</a> (const vector&lt; int &gt; &amp;genotype)</td></tr>
<tr class="separator:a5fc66cb05c4ba80e39d078af9ddce715"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a55c456d612728331007f2f6c4307195f"><td class="memItemLeft" align="right" valign="top">vector&lt; vcflib::Variant &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a55c456d612728331007f2f6c4307195f">locus_to_variant</a> (<a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph, const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *snarl, const pair&lt; unordered_set&lt; <a class="el" href="namespacevg.html#afaa15fd8bdfd688d48561e1b12aee821">id_t</a> &gt;, unordered_set&lt; <a class="el" href="namespacevg.html#a3de8fe5f38b1730a36af4ad9c4258963">edge_t</a> &gt; &gt; &amp;contents, const <a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a> &amp;manager, const <a class="el" href="structvg_1_1PathIndex.html">PathIndex</a> &amp;index, vcflib::VariantCallFile &amp;vcf, const <a class="el" href="structvg_1_1Locus.html">Locus</a> &amp;locus, const string &amp;sample_name=&quot;SAMPLE&quot;)</td></tr>
<tr class="separator:a55c456d612728331007f2f6c4307195f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad2beb83f750d1deec75b16a60c2ea805"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#ad2beb83f750d1deec75b16a60c2ea805">write_vcf_header</a> (std::ostream &amp;stream, const std::string &amp;sample_name, const std::string &amp;contig_name, size_t contig_size)</td></tr>
<tr class="separator:ad2beb83f750d1deec75b16a60c2ea805"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8a5aed6de86a488fe78a223d5e4d6bd6"><td class="memItemLeft" align="right" valign="top">vcflib::VariantCallFile *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a8a5aed6de86a488fe78a223d5e4d6bd6">start_vcf</a> (std::ostream &amp;stream, const <a class="el" href="structvg_1_1PathIndex.html">PathIndex</a> &amp;index, const string &amp;sample_name, const string &amp;contig_name, size_t contig_size)</td></tr>
<tr class="separator:a8a5aed6de86a488fe78a223d5e4d6bd6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a29a6f2fd20161c9313498760578c091a"><td class="memItemLeft" align="right" valign="top">pair&lt; pair&lt; int64_t, int64_t &gt;, bool &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a29a6f2fd20161c9313498760578c091a">get_snarl_reference_bounds</a> (const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *snarl, const <a class="el" href="structvg_1_1PathIndex.html">PathIndex</a> &amp;index, const <a class="el" href="classhandlegraph_1_1HandleGraph.html">HandleGraph</a> *graph)</td></tr>
<tr class="separator:a29a6f2fd20161c9313498760578c091a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6c9c525615b0c453f0685ea2ed987cde"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a6c9c525615b0c453f0685ea2ed987cde">report_snarl</a> (const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *snarl, const <a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a> &amp;manager, const <a class="el" href="structvg_1_1PathIndex.html">PathIndex</a> *index, <a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph, <a class="el" href="structvg_1_1PathIndex.html">PathIndex</a> *reference_index)</td></tr>
<tr class="separator:a6c9c525615b0c453f0685ea2ed987cde"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1a141641a0939e35dc6ee19281001628"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a1a141641a0939e35dc6ee19281001628">report_snarl_traversal</a> (const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *snarl, const <a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a> &amp;manager, <a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph)</td></tr>
<tr class="separator:a1a141641a0939e35dc6ee19281001628"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abe8c49eaf4799ae7202e19b6e09288a7"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#abe8c49eaf4799ae7202e19b6e09288a7">report_affinities</a> (map&lt; const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> *, vector&lt; <a class="el" href="structvg_1_1Genotyper_1_1Affinity.html">Genotyper::Affinity</a> &gt;&gt; &amp;affinities, vector&lt; <a class="el" href="structvg_1_1SnarlTraversal.html">SnarlTraversal</a> &gt; &amp;paths, <a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph)</td></tr>
<tr class="separator:abe8c49eaf4799ae7202e19b6e09288a7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af5b580b6db4e727934f964d00b058038"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#af5b580b6db4e727934f964d00b058038">write_variant</a> ()</td></tr>
<tr class="separator:af5b580b6db4e727934f964d00b058038"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aadb967b3ab9daa067f8451c29fda0a5e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#aadb967b3ab9daa067f8451c29fda0a5e">print_statistics</a> (ostream &amp;out)</td></tr>
<tr class="separator:aadb967b3ab9daa067f8451c29fda0a5e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab11a87f5f21d3a7122ec85f6d853f336"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classvg_1_1VG.html">VG</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#ab11a87f5f21d3a7122ec85f6d853f336">make_subset_graph</a> (<a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph, const string &amp;ref_path_name, map&lt; string, const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> * &gt; &amp;reads_by_name)</td></tr>
<tr class="separator:ab11a87f5f21d3a7122ec85f6d853f336"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aae52b974044843aed34fc31152ccda77"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#aae52b974044843aed34fc31152ccda77">edge_allele_labels</a> (const <a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph, const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *snarl, const vector&lt; list&lt; <a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a> &gt;&gt; &amp;superbubble_paths, unordered_map&lt; pair&lt; <a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a>, <a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a> &gt;, unordered_set&lt; size_t &gt;, <a class="el" href="structvg_1_1Genotyper_1_1hash__oriented__edge.html">hash_oriented_edge</a> &gt; *out_edge_allele_sets)</td></tr>
<tr class="separator:aae52b974044843aed34fc31152ccda77"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa46523cd1493686b28210136d9baf713"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#aa46523cd1493686b28210136d9baf713">allele_ambiguity_log_probs</a> (const <a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph, const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *snarl, const vector&lt; list&lt; <a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a> &gt;&gt; &amp;superbubble_paths, const unordered_map&lt; pair&lt; <a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a>, <a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a> &gt;, unordered_set&lt; size_t &gt;, <a class="el" href="structvg_1_1Genotyper_1_1hash__oriented__edge.html">hash_oriented_edge</a> &gt; &amp;edge_allele_sets, vector&lt; unordered_map&lt; vector&lt; size_t &gt;, double, <a class="el" href="structvg_1_1Genotyper_1_1hash__ambiguous__allele__set.html">hash_ambiguous_allele_set</a> &gt;&gt; *out_allele_ambiguity_probs)</td></tr>
<tr class="separator:aa46523cd1493686b28210136d9baf713"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:aa2b4d94f59265c107895c48a26b107a6"><td class="memItemLeft" align="right" valign="top">static bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#aa2b4d94f59265c107895c48a26b107a6">mapping_enters_side</a> (const <a class="el" href="structvg_1_1Mapping.html">Mapping</a> &amp;mapping, const <a class="el" href="structhandlegraph_1_1handle__t.html">handle_t</a> &amp;side, const <a class="el" href="classhandlegraph_1_1HandleGraph.html">HandleGraph</a> *graph)</td></tr>
<tr class="separator:aa2b4d94f59265c107895c48a26b107a6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8b054a94fb2ceb4c65731188227b0cf8"><td class="memItemLeft" align="right" valign="top">static bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a8b054a94fb2ceb4c65731188227b0cf8">mapping_exits_side</a> (const <a class="el" href="structvg_1_1Mapping.html">Mapping</a> &amp;mapping, const <a class="el" href="structhandlegraph_1_1handle__t.html">handle_t</a> &amp;side, const <a class="el" href="classhandlegraph_1_1HandleGraph.html">HandleGraph</a> *graph)</td></tr>
<tr class="separator:a8b054a94fb2ceb4c65731188227b0cf8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7c7f091482612f6b0cd2baf0c19c45fb"><td class="memItemLeft" align="right" valign="top">static bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a7c7f091482612f6b0cd2baf0c19c45fb">is_snarl_smaller_than_reads</a> (<a class="el" href="structvg_1_1AugmentedGraph.html">AugmentedGraph</a> &amp;augmented_graph, const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *snarl, const pair&lt; unordered_set&lt; <a class="el" href="namespacevg.html#afaa15fd8bdfd688d48561e1b12aee821">id_t</a> &gt;, unordered_set&lt; <a class="el" href="namespacevg.html#a3de8fe5f38b1730a36af4ad9c4258963">edge_t</a> &gt; &gt; &amp;contents, map&lt; string, const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> * &gt; &amp;reads_by_name)</td></tr>
<tr class="separator:a7c7f091482612f6b0cd2baf0c19c45fb"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:a500af110546f9572fe40d5ef098552d9"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a500af110546f9572fe40d5ef098552d9">max_path_search_steps</a> = 100</td></tr>
<tr class="separator:a500af110546f9572fe40d5ef098552d9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab3e4270419b361c9455125ca305709e2"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#ab3e4270419b361c9455125ca305709e2">unfold_max_length</a> = 200</td></tr>
<tr class="separator:ab3e4270419b361c9455125ca305709e2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a384f22f52c2a1f78b7ca613468e2e186"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a384f22f52c2a1f78b7ca613468e2e186">dagify_steps</a> = 1</td></tr>
<tr class="separator:a384f22f52c2a1f78b7ca613468e2e186"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a96285d96b960d50199ad3167c20ddf91"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a96285d96b960d50199ad3167c20ddf91">max_het_bias</a> = 4</td></tr>
<tr class="separator:a96285d96b960d50199ad3167c20ddf91"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5991ce66cc2a9ea45bd27d288dd7d9b8"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a5991ce66cc2a9ea45bd27d288dd7d9b8">use_mapq</a> = false</td></tr>
<tr class="separator:a5991ce66cc2a9ea45bd27d288dd7d9b8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a79ac39be882405a426afaeeef65c9660"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a79ac39be882405a426afaeeef65c9660">realign_indels</a> = false</td></tr>
<tr class="separator:a79ac39be882405a426afaeeef65c9660"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3c40092b41694cacc7e531bf7f9fb8aa"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a3c40092b41694cacc7e531bf7f9fb8aa">default_sequence_quality</a> = 15</td></tr>
<tr class="separator:a3c40092b41694cacc7e531bf7f9fb8aa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afffa8d75fd4155bc7d420a41d92f198c"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#afffa8d75fd4155bc7d420a41d92f198c">min_recurrence</a> = 2</td></tr>
<tr class="separator:afffa8d75fd4155bc7d420a41d92f198c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6b1feebb10944e684e417115e880fe39"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a6b1feebb10944e684e417115e880fe39">min_unique_per_strand</a> = 2</td></tr>
<tr class="separator:a6b1feebb10944e684e417115e880fe39"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8fc1fdccfa9829bb29903ed5d7b21b28"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a8fc1fdccfa9829bb29903ed5d7b21b28">min_score_per_base</a> = 0.90</td></tr>
<tr class="separator:a8fc1fdccfa9829bb29903ed5d7b21b28"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac748204c3fb19573569172ee57b1d082"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#ac748204c3fb19573569172ee57b1d082">diploid_prior_logprob</a> = <a class="el" href="namespacevg.html#a476a2102f4f5aa82c06be87780ae6d42">prob_to_logprob</a>(0.999)</td></tr>
<tr class="separator:ac748204c3fb19573569172ee57b1d082"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a21799af8d6da86a3bbcfffb1060d27cf"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a21799af8d6da86a3bbcfffb1060d27cf">het_prior_logprob</a> = <a class="el" href="namespacevg.html#a476a2102f4f5aa82c06be87780ae6d42">prob_to_logprob</a>(0.1)</td></tr>
<tr class="separator:a21799af8d6da86a3bbcfffb1060d27cf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a76e36b8cc1f5a8a5522c5ae3f57b5464"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a76e36b8cc1f5a8a5522c5ae3f57b5464">haploid_prior_logprob</a> = <a class="el" href="namespacevg.html#a476a2102f4f5aa82c06be87780ae6d42">prob_to_logprob</a>(0.5)</td></tr>
<tr class="separator:a76e36b8cc1f5a8a5522c5ae3f57b5464"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a853fb8840552901b359153ec80aa81e1"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a853fb8840552901b359153ec80aa81e1">deleted_prior_logprob</a> = <a class="el" href="namespacevg.html#a476a2102f4f5aa82c06be87780ae6d42">prob_to_logprob</a>(0.1)</td></tr>
<tr class="separator:a853fb8840552901b359153ec80aa81e1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a97aef0fe84f0514e17e2ec8e00162225"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a97aef0fe84f0514e17e2ec8e00162225">polyploid_prior_success_logprob</a> = <a class="el" href="namespacevg.html#a476a2102f4f5aa82c06be87780ae6d42">prob_to_logprob</a>(0.5)</td></tr>
<tr class="separator:a97aef0fe84f0514e17e2ec8e00162225"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad14928554aace83524e80d8c109e17b2"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classvg_1_1Translator.html">Translator</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#ad14928554aace83524e80d8c109e17b2">translator</a></td></tr>
<tr class="separator:ad14928554aace83524e80d8c109e17b2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac0d4917131823fc2dab43927347d4623"><td class="memItemLeft" align="right" valign="top">unordered_map&lt; size_t, size_t &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#ac0d4917131823fc2dab43927347d4623">snarl_reference_length_histogram</a></td></tr>
<tr class="separator:ac0d4917131823fc2dab43927347d4623"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7ec1915674299d727c61377e91385974"><td class="memItemLeft" align="right" valign="top">unordered_set&lt; const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> * &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a7ec1915674299d727c61377e91385974">snarl_traversals</a></td></tr>
<tr class="separator:a7ec1915674299d727c61377e91385974"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac177e2664adf7be9b92939148ea77c65"><td class="memItemLeft" align="right" valign="top">unordered_set&lt; const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> * &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#ac177e2664adf7be9b92939148ea77c65">all_snarls</a></td></tr>
<tr class="separator:ac177e2664adf7be9b92939148ea77c65"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a91d6174a20ca4f121f68766fbf7fb4b0"><td class="memItemLeft" align="right" valign="top">vector&lt; <a class="el" href="classvg_1_1Aligner.html">Aligner</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a91d6174a20ca4f121f68766fbf7fb4b0">normal_aligners</a></td></tr>
<tr class="separator:a91d6174a20ca4f121f68766fbf7fb4b0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2705913bdfb5f6679afb5475a22ecf61"><td class="memItemLeft" align="right" valign="top">vector&lt; <a class="el" href="classvg_1_1QualAdjAligner.html">QualAdjAligner</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a2705913bdfb5f6679afb5475a22ecf61">quality_aligners</a></td></tr>
<tr class="separator:a2705913bdfb5f6679afb5475a22ecf61"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adc250029dd72b833de1042e972b2a8d7"><td class="memItemLeft" align="right" valign="top">map&lt; <a class="el" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4">TraversalAlg</a>, string &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#adc250029dd72b833de1042e972b2a8d7">alg2name</a></td></tr>
<tr class="separator:adc250029dd72b833de1042e972b2a8d7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5b040c1636f479b52dbdb641e82f74dd"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4">TraversalAlg</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a5b040c1636f479b52dbdb641e82f74dd">traversal_alg</a> = TraversalAlg::Reads</td></tr>
<tr class="separator:a5b040c1636f479b52dbdb641e82f74dd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a72599c00acaed553e0d235e5c73d4b82"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Genotyper.html#a72599c00acaed553e0d235e5c73d4b82">show_progress</a> = false</td></tr>
<tr class="separator:a72599c00acaed553e0d235e5c73d4b82"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Class to hold on to genotyping parameters and genotyping functions. </p>
</div><h2 class="groupheader">Member Enumeration Documentation</h2>
<a id="a6e4dce6cd9cc24ab6857ceeaed6db3b4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6e4dce6cd9cc24ab6857ceeaed6db3b4">&#9670;&nbsp;</a></span>TraversalAlg</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">enum <a class="el" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4">vg::Genotyper::TraversalAlg</a></td>
        </tr>
      </table>
</div><div class="memdoc">
<table class="fieldtable">
<tr><th colspan="2">Enumerator</th></tr><tr><td class="fieldname"><a id="a6e4dce6cd9cc24ab6857ceeaed6db3b4a9989ae14c6b6c27998126ab745488926"></a>Reads&#160;</td><td class="fielddoc"></td></tr>
<tr><td class="fieldname"><a id="a6e4dce6cd9cc24ab6857ceeaed6db3b4a03b965f4f4d03d6869811e0dedbfb0dd"></a>Exhaustive&#160;</td><td class="fielddoc"></td></tr>
<tr><td class="fieldname"><a id="a6e4dce6cd9cc24ab6857ceeaed6db3b4aca000fa2f53ba37764dedd204e1fd27e"></a>Representative&#160;</td><td class="fielddoc"></td></tr>
<tr><td class="fieldname"><a id="a6e4dce6cd9cc24ab6857ceeaed6db3b4ab530b2114d4cd7f36bb3a41c01928995"></a>Adaptive&#160;</td><td class="fielddoc"></td></tr>
</table>

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="a76bba1da02c9a75895896efd5b5e3641"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a76bba1da02c9a75895896efd5b5e3641">&#9670;&nbsp;</a></span>alignment_qual_score()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vg::Genotyper::alignment_qual_score </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *&#160;</td>
          <td class="paramname"><em>snarl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> &amp;&#160;</td>
          <td class="paramname"><em>alignment</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Given an <a class="el" href="structvg_1_1Alignment.html">Alignment</a> and a <a class="el" href="structvg_1_1Snarl.html" title="Describes a subgraph that is connected to the rest of the graph by two nodes.">Snarl</a>, compute a phred score for the quality of the alignment's bases within the snarl overall (not counting the start and end nodes), which is supposed to be interpretable as the probability that the call of the sequence is wrong (to the degree that it would no longer support the alleles it appears to support).</p>
<p>In practice we're just going to average the quality scores for all the bases interior to the snarl (i.e. not counting the start and end nodes).</p>
<p>If the alignment doesn't have base qualities, or no qualities are available for bases internal to the snarl, returns a default value. </p>

</div>
</div>
<a id="aa46523cd1493686b28210136d9baf713"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa46523cd1493686b28210136d9baf713">&#9670;&nbsp;</a></span>allele_ambiguity_log_probs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Genotyper::allele_ambiguity_log_probs </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *&#160;</td>
          <td class="paramname"><em>snarl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const vector&lt; list&lt; <a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a> &gt;&gt; &amp;&#160;</td>
          <td class="paramname"><em>superbubble_paths</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const unordered_map&lt; pair&lt; <a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a>, <a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a> &gt;, unordered_set&lt; size_t &gt;, <a class="el" href="structvg_1_1Genotyper_1_1hash__oriented__edge.html">hash_oriented_edge</a> &gt; &amp;&#160;</td>
          <td class="paramname"><em>edge_allele_sets</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">vector&lt; unordered_map&lt; vector&lt; size_t &gt;, double, <a class="el" href="structvg_1_1Genotyper_1_1hash__ambiguous__allele__set.html">hash_ambiguous_allele_set</a> &gt;&gt; *&#160;</td>
          <td class="paramname"><em>out_allele_ambiguity_probs</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="aae52b974044843aed34fc31152ccda77"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aae52b974044843aed34fc31152ccda77">&#9670;&nbsp;</a></span>edge_allele_labels()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Genotyper::edge_allele_labels </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *&#160;</td>
          <td class="paramname"><em>snarl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const vector&lt; list&lt; <a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a> &gt;&gt; &amp;&#160;</td>
          <td class="paramname"><em>superbubble_paths</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unordered_map&lt; pair&lt; <a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a>, <a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a> &gt;, unordered_set&lt; size_t &gt;, <a class="el" href="structvg_1_1Genotyper_1_1hash__oriented__edge.html">hash_oriented_edge</a> &gt; *&#160;</td>
          <td class="paramname"><em>out_edge_allele_sets</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="ae6a3a060adcffc32724498bbfda2c771"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae6a3a060adcffc32724498bbfda2c771">&#9670;&nbsp;</a></span>genotype_snarl()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structvg_1_1Locus.html">Locus</a> vg::Genotyper::genotype_snarl </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *&#160;</td>
          <td class="paramname"><em>snarl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const vector&lt; <a class="el" href="structvg_1_1SnarlTraversal.html">SnarlTraversal</a> &gt; &amp;&#160;</td>
          <td class="paramname"><em>superbubble_paths</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const map&lt; const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> *, vector&lt; <a class="el" href="structvg_1_1Genotyper_1_1Affinity.html">Affinity</a> &gt;&gt; &amp;&#160;</td>
          <td class="paramname"><em>affinities</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Compute annotated genotype from affinities and superbubble paths. Needs access to the graph so it can chop up the alignments, which requires node sizes. </p>

</div>
</div>
<a id="abea795c1969c4866c349a07c7f9c7102"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abea795c1969c4866c349a07c7f9c7102">&#9670;&nbsp;</a></span>get_affinities()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">map&lt; const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> *, vector&lt; <a class="el" href="structvg_1_1Genotyper_1_1Affinity.html">Genotyper::Affinity</a> &gt; &gt; vg::Genotyper::get_affinities </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structvg_1_1AugmentedGraph.html">AugmentedGraph</a> &amp;&#160;</td>
          <td class="paramname"><em>aug</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const map&lt; string, const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> * &gt; &amp;&#160;</td>
          <td class="paramname"><em>reads_by_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *&#160;</td>
          <td class="paramname"><em>snarl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const pair&lt; unordered_set&lt; <a class="el" href="namespacevg.html#afaa15fd8bdfd688d48561e1b12aee821">id_t</a> &gt;, unordered_set&lt; <a class="el" href="namespacevg.html#a3de8fe5f38b1730a36af4ad9c4258963">edge_t</a> &gt; &gt; &amp;&#160;</td>
          <td class="paramname"><em>contents</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a> &amp;&#160;</td>
          <td class="paramname"><em>manager</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const vector&lt; <a class="el" href="structvg_1_1SnarlTraversal.html">SnarlTraversal</a> &gt; &amp;&#160;</td>
          <td class="paramname"><em>superbubble_paths</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get the affinity of all the reads relevant to the superbubble to all the paths through the superbubble.</p>
<p><a class="el" href="structvg_1_1Genotyper_1_1Affinity.html">Affinity</a> is a double out of 1.0. Higher is better. </p>

</div>
</div>
<a id="a966eee3f764281271904996546ef205b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a966eee3f764281271904996546ef205b">&#9670;&nbsp;</a></span>get_affinities_fast()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">map&lt; const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> *, vector&lt; <a class="el" href="structvg_1_1Genotyper_1_1Affinity.html">Genotyper::Affinity</a> &gt; &gt; vg::Genotyper::get_affinities_fast </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structvg_1_1AugmentedGraph.html">AugmentedGraph</a> &amp;&#160;</td>
          <td class="paramname"><em>aug</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const map&lt; string, const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> * &gt; &amp;&#160;</td>
          <td class="paramname"><em>reads_by_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *&#160;</td>
          <td class="paramname"><em>snarl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const pair&lt; unordered_set&lt; <a class="el" href="namespacevg.html#afaa15fd8bdfd688d48561e1b12aee821">id_t</a> &gt;, unordered_set&lt; <a class="el" href="namespacevg.html#a3de8fe5f38b1730a36af4ad9c4258963">edge_t</a> &gt; &gt; &amp;&#160;</td>
          <td class="paramname"><em>contents</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a> &amp;&#160;</td>
          <td class="paramname"><em>manager</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const vector&lt; <a class="el" href="structvg_1_1SnarlTraversal.html">SnarlTraversal</a> &gt; &amp;&#160;</td>
          <td class="paramname"><em>superbubble_paths</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>allow_internal_alignments</em> = <code>false</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get affinities as above but using only string comparison instead of alignment. Affinities are 0 for mismatch and 1 for a perfect match. </p>

</div>
</div>
<a id="a9071cdf5ae591b478feaac025b069772"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9071cdf5ae591b478feaac025b069772">&#9670;&nbsp;</a></span>get_genotype_log_likelihood()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vg::Genotyper::get_genotype_log_likelihood </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *&#160;</td>
          <td class="paramname"><em>snarl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const vector&lt; int &gt; &amp;&#160;</td>
          <td class="paramname"><em>genotype</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const vector&lt; pair&lt; const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> *, vector&lt; <a class="el" href="structvg_1_1Genotyper_1_1Affinity.html">Affinity</a> &gt;&gt;&gt; &amp;&#160;</td>
          <td class="paramname"><em>alignment_consistency</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Compute the probability of the observed alignments given the genotype.</p>
<p>Takes a genotype as a vector of allele numbers, and support data as a collection of pairs of Alignments and vectors of bools marking whether each alignment is consistent with each allele.</p>
<p>Alignments should have had their quality values trimmed down to just the part covering the superbubble.</p>
<p>Returns a natural log likelihood. </p>

</div>
</div>
<a id="a5fc66cb05c4ba80e39d078af9ddce715"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5fc66cb05c4ba80e39d078af9ddce715">&#9670;&nbsp;</a></span>get_genotype_log_prior()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vg::Genotyper::get_genotype_log_prior </td>
          <td>(</td>
          <td class="paramtype">const vector&lt; int &gt; &amp;&#160;</td>
          <td class="paramname"><em>genotype</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Compute the prior probability of the given genotype.</p>
<p>Takes a genotype as a vector of allele numbers. It is not guaranteed that allele 0 corresponds to any notion of primary reference-ness.</p>
<p>Returns a natural log prior probability.</p>
<p>TODO: add in strand bias </p>

</div>
</div>
<a id="af6518141a6e6f122e5f534c041d16918"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af6518141a6e6f122e5f534c041d16918">&#9670;&nbsp;</a></span>get_qualities_in_snarl()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">string vg::Genotyper::get_qualities_in_snarl </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *&#160;</td>
          <td class="paramname"><em>snarl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> &amp;&#160;</td>
          <td class="paramname"><em>alignment</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get all the quality values in the alignment between the start and end nodes of a snarl. Handles alignments that enter the snarl from the end, and alignments that never make it through the snarl.</p>
<p>If we run out of qualities, or qualities aren't present, returns no qualities.</p>
<p>If an alignment goes through the snarl multipe times, we get all the qualities from when it is in the snarl.</p>
<p>Does not return qualities on the start and end nodes. May return an empty string. </p>

</div>
</div>
<a id="a29a6f2fd20161c9313498760578c091a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a29a6f2fd20161c9313498760578c091a">&#9670;&nbsp;</a></span>get_snarl_reference_bounds()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">pair&lt; pair&lt; int64_t, int64_t &gt;, bool &gt; vg::Genotyper::get_snarl_reference_bounds </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *&#160;</td>
          <td class="paramname"><em>snarl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1PathIndex.html">PathIndex</a> &amp;&#160;</td>
          <td class="paramname"><em>index</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classhandlegraph_1_1HandleGraph.html">HandleGraph</a> *&#160;</td>
          <td class="paramname"><em>graph</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Utility function for getting the reference bounds (start and past-end) of a snarl with relation to a given reference index in the given graph. Computes bounds of the variable region, not including the fixed start and end node lengths. Also returns whether the reference path goes through the snarl forwards (false) or backwards (true). </p>

</div>
</div>
<a id="a7a14f1e259fe612b944aadf701ca1585"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7a14f1e259fe612b944aadf701ca1585">&#9670;&nbsp;</a></span>get_snarl_traversals()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt; <a class="el" href="structvg_1_1SnarlTraversal.html">SnarlTraversal</a> &gt; vg::Genotyper::get_snarl_traversals </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structvg_1_1AugmentedGraph.html">AugmentedGraph</a> &amp;&#160;</td>
          <td class="paramname"><em>augmented_graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a> &amp;&#160;</td>
          <td class="paramname"><em>manager</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">map&lt; string, const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> * &gt; &amp;&#160;</td>
          <td class="paramname"><em>reads_by_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *&#160;</td>
          <td class="paramname"><em>snarl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const pair&lt; unordered_set&lt; <a class="el" href="namespacevg.html#afaa15fd8bdfd688d48561e1b12aee821">id_t</a> &gt;, unordered_set&lt; <a class="el" href="namespacevg.html#a3de8fe5f38b1730a36af4ad9c4258963">edge_t</a> &gt; &gt; &amp;&#160;</td>
          <td class="paramname"><em>contents</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structvg_1_1PathIndex.html">PathIndex</a> *&#160;</td>
          <td class="paramname"><em>reference_index</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4">TraversalAlg</a>&#160;</td>
          <td class="paramname"><em>use_traversal_alg</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Get traversals of a snarl in one of several ways. </p>

</div>
</div>
<a id="a7c7f091482612f6b0cd2baf0c19c45fb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7c7f091482612f6b0cd2baf0c19c45fb">&#9670;&nbsp;</a></span>is_snarl_smaller_than_reads()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Genotyper::is_snarl_smaller_than_reads </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structvg_1_1AugmentedGraph.html">AugmentedGraph</a> &amp;&#160;</td>
          <td class="paramname"><em>augmented_graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *&#160;</td>
          <td class="paramname"><em>snarl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const pair&lt; unordered_set&lt; <a class="el" href="namespacevg.html#afaa15fd8bdfd688d48561e1b12aee821">id_t</a> &gt;, unordered_set&lt; <a class="el" href="namespacevg.html#a3de8fe5f38b1730a36af4ad9c4258963">edge_t</a> &gt; &gt; &amp;&#160;</td>
          <td class="paramname"><em>contents</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">map&lt; string, const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> * &gt; &amp;&#160;</td>
          <td class="paramname"><em>reads_by_name</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Check if a snarl is small enough to be covered by reads (very conservative) </p>

</div>
</div>
<a id="a55c456d612728331007f2f6c4307195f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a55c456d612728331007f2f6c4307195f">&#9670;&nbsp;</a></span>locus_to_variant()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt; vcflib::Variant &gt; vg::Genotyper::locus_to_variant </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *&#160;</td>
          <td class="paramname"><em>snarl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const pair&lt; unordered_set&lt; <a class="el" href="namespacevg.html#afaa15fd8bdfd688d48561e1b12aee821">id_t</a> &gt;, unordered_set&lt; <a class="el" href="namespacevg.html#a3de8fe5f38b1730a36af4ad9c4258963">edge_t</a> &gt; &gt; &amp;&#160;</td>
          <td class="paramname"><em>contents</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a> &amp;&#160;</td>
          <td class="paramname"><em>manager</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1PathIndex.html">PathIndex</a> &amp;&#160;</td>
          <td class="paramname"><em>index</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">vcflib::VariantCallFile &amp;&#160;</td>
          <td class="paramname"><em>vcf</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Locus.html">Locus</a> &amp;&#160;</td>
          <td class="paramname"><em>locus</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>sample_name</em> = <code>&quot;SAMPLE&quot;</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Make a VCFlib variant from a called <a class="el" href="structvg_1_1Locus.html" title="Describes a genetic locus with multiple possible alleles, a genotype, and observational support.">Locus</a>. Depends on an index of the reference path we want to call against.</p>
<p>Returns 0 or more variants we can articulate from the superbubble. Sometimes if we can't make a variant for the superbubble against the reference path, we'll emit 0 variants. </p>

</div>
</div>
<a id="ab11a87f5f21d3a7122ec85f6d853f336"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab11a87f5f21d3a7122ec85f6d853f336">&#9670;&nbsp;</a></span>make_subset_graph()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classvg_1_1VG.html">VG</a> vg::Genotyper::make_subset_graph </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>ref_path_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">map&lt; string, const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> * &gt; &amp;&#160;</td>
          <td class="paramname"><em>reads_by_name</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Subset the graph </p>

</div>
</div>
<a id="aa2b4d94f59265c107895c48a26b107a6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa2b4d94f59265c107895c48a26b107a6">&#9670;&nbsp;</a></span>mapping_enters_side()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static bool vg::Genotyper::mapping_enters_side </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Mapping.html">Mapping</a> &amp;&#160;</td>
          <td class="paramname"><em>mapping</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structhandlegraph_1_1handle__t.html">handle_t</a> &amp;&#160;</td>
          <td class="paramname"><em>side</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classhandlegraph_1_1HandleGraph.html">HandleGraph</a> *&#160;</td>
          <td class="paramname"><em>graph</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Check if a mapping corresponds to the beginning or end of snarl by making sure it crosses the given side in the expected direction. The handle should be forward for the left side and reverse for the right side. </p>

</div>
</div>
<a id="a8b054a94fb2ceb4c65731188227b0cf8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8b054a94fb2ceb4c65731188227b0cf8">&#9670;&nbsp;</a></span>mapping_exits_side()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static bool vg::Genotyper::mapping_exits_side </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Mapping.html">Mapping</a> &amp;&#160;</td>
          <td class="paramname"><em>mapping</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structhandlegraph_1_1handle__t.html">handle_t</a> &amp;&#160;</td>
          <td class="paramname"><em>side</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classhandlegraph_1_1HandleGraph.html">HandleGraph</a> *&#160;</td>
          <td class="paramname"><em>graph</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">static</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a id="aadb967b3ab9daa067f8451c29fda0a5e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aadb967b3ab9daa067f8451c29fda0a5e">&#9670;&nbsp;</a></span>print_statistics()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Genotyper::print_statistics </td>
          <td>(</td>
          <td class="paramtype">ostream &amp;&#160;</td>
          <td class="paramname"><em>out</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Print snarl statistics to the given stream. </p>

</div>
</div>
<a id="abe8c49eaf4799ae7202e19b6e09288a7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abe8c49eaf4799ae7202e19b6e09288a7">&#9670;&nbsp;</a></span>report_affinities()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Genotyper::report_affinities </td>
          <td>(</td>
          <td class="paramtype">map&lt; const <a class="el" href="structvg_1_1Alignment.html">Alignment</a> *, vector&lt; <a class="el" href="structvg_1_1Genotyper_1_1Affinity.html">Genotyper::Affinity</a> &gt;&gt; &amp;&#160;</td>
          <td class="paramname"><em>affinities</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">vector&lt; <a class="el" href="structvg_1_1SnarlTraversal.html">SnarlTraversal</a> &gt; &amp;&#160;</td>
          <td class="paramname"><em>paths</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Print some information about affinities </p>

</div>
</div>
<a id="a6c9c525615b0c453f0685ea2ed987cde"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6c9c525615b0c453f0685ea2ed987cde">&#9670;&nbsp;</a></span>report_snarl()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Genotyper::report_snarl </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *&#160;</td>
          <td class="paramname"><em>snarl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a> &amp;&#160;</td>
          <td class="paramname"><em>manager</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1PathIndex.html">PathIndex</a> *&#160;</td>
          <td class="paramname"><em>index</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structvg_1_1PathIndex.html">PathIndex</a> *&#160;</td>
          <td class="paramname"><em>reference_index</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Tell the statistics tracking code that a snarl exists. We can do things like count up the snarl length in the reference and so on. Called only once per snarl, but may be called on multiple threads simultaneously. </p>

</div>
</div>
<a id="a1a141641a0939e35dc6ee19281001628"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1a141641a0939e35dc6ee19281001628">&#9670;&nbsp;</a></span>report_snarl_traversal()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Genotyper::report_snarl_traversal </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *&#160;</td>
          <td class="paramname"><em>snarl</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a> &amp;&#160;</td>
          <td class="paramname"><em>manager</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Tell the statistics tracking code that a read traverses a snarl completely. May be called multiple times for a given read and snarl, and may be called in parallel. </p>

</div>
</div>
<a id="af760be5c99e086b2c41510b3d3a47e5a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af760be5c99e086b2c41510b3d3a47e5a">&#9670;&nbsp;</a></span>run()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Genotyper::run </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structvg_1_1AugmentedGraph.html">AugmentedGraph</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">ostream &amp;&#160;</td>
          <td class="paramname"><em>out</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">string&#160;</td>
          <td class="paramname"><em>ref_path_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">string&#160;</td>
          <td class="paramname"><em>contig_name</em> = <code>&quot;&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">string&#160;</td>
          <td class="paramname"><em>sample_name</em> = <code>&quot;&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">string&#160;</td>
          <td class="paramname"><em>augmented_file_name</em> = <code>&quot;&quot;</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>output_vcf</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">bool&#160;</td>
          <td class="paramname"><em>output_json</em> = <code>false</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>length_override</em> = <code>0</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>variant_offset</em> = <code>0</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Process and write output. Alignments must be embedded in the <a class="el" href="structvg_1_1AugmentedGraph.html">AugmentedGraph</a>. </p>

</div>
</div>
<a id="a8a5aed6de86a488fe78a223d5e4d6bd6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8a5aed6de86a488fe78a223d5e4d6bd6">&#9670;&nbsp;</a></span>start_vcf()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vcflib::VariantCallFile * vg::Genotyper::start_vcf </td>
          <td>(</td>
          <td class="paramtype">std::ostream &amp;&#160;</td>
          <td class="paramname"><em>stream</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1PathIndex.html">PathIndex</a> &amp;&#160;</td>
          <td class="paramname"><em>index</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>sample_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const string &amp;&#160;</td>
          <td class="paramname"><em>contig_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>contig_size</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Start VCF output to a stream. Returns a VCFlib VariantCallFile that needs to be deleted. </p>

</div>
</div>
<a id="af5b580b6db4e727934f964d00b058038"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af5b580b6db4e727934f964d00b058038">&#9670;&nbsp;</a></span>write_variant()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Genotyper::write_variant </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Write the variant as output </p>

</div>
</div>
<a id="ad2beb83f750d1deec75b16a60c2ea805"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad2beb83f750d1deec75b16a60c2ea805">&#9670;&nbsp;</a></span>write_vcf_header()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Genotyper::write_vcf_header </td>
          <td>(</td>
          <td class="paramtype">std::ostream &amp;&#160;</td>
          <td class="paramname"><em>stream</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>sample_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const std::string &amp;&#160;</td>
          <td class="paramname"><em>contig_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>contig_size</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Make a VCF header </p>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a id="adc250029dd72b833de1042e972b2a8d7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#adc250029dd72b833de1042e972b2a8d7">&#9670;&nbsp;</a></span>alg2name</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">map&lt;<a class="el" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4">TraversalAlg</a>, string&gt; vg::Genotyper::alg2name</td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Initial value:</b><div class="fragment"><div class="line">= {</div>
<div class="line">        {<a class="code" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4a9989ae14c6b6c27998126ab745488926">Reads</a>, <span class="stringliteral">&quot;Reads&quot;</span>},</div>
<div class="line">        {<a class="code" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4a03b965f4f4d03d6869811e0dedbfb0dd">Exhaustive</a>, <span class="stringliteral">&quot;Exhaustive&quot;</span>},</div>
<div class="line">        {<a class="code" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4aca000fa2f53ba37764dedd204e1fd27e">Representative</a>, <span class="stringliteral">&quot;Representative&quot;</span>},</div>
<div class="line">        {<a class="code" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4ab530b2114d4cd7f36bb3a41c01928995">Adaptive</a>, <span class="stringliteral">&quot;Adaptive&quot;</span>}</div>
<div class="line">    }</div>
<div class="ttc" id="aclassvg_1_1Genotyper_html_a6e4dce6cd9cc24ab6857ceeaed6db3b4a03b965f4f4d03d6869811e0dedbfb0dd"><div class="ttname"><a href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4a03b965f4f4d03d6869811e0dedbfb0dd">vg::Genotyper::Exhaustive</a></div><div class="ttdeci">@ Exhaustive</div><div class="ttdef"><b>Definition:</b> genotyper.hpp:140</div></div>
<div class="ttc" id="aclassvg_1_1Genotyper_html_a6e4dce6cd9cc24ab6857ceeaed6db3b4a9989ae14c6b6c27998126ab745488926"><div class="ttname"><a href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4a9989ae14c6b6c27998126ab745488926">vg::Genotyper::Reads</a></div><div class="ttdeci">@ Reads</div><div class="ttdef"><b>Definition:</b> genotyper.hpp:140</div></div>
<div class="ttc" id="aclassvg_1_1Genotyper_html_a6e4dce6cd9cc24ab6857ceeaed6db3b4ab530b2114d4cd7f36bb3a41c01928995"><div class="ttname"><a href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4ab530b2114d4cd7f36bb3a41c01928995">vg::Genotyper::Adaptive</a></div><div class="ttdeci">@ Adaptive</div><div class="ttdef"><b>Definition:</b> genotyper.hpp:140</div></div>
<div class="ttc" id="aclassvg_1_1Genotyper_html_a6e4dce6cd9cc24ab6857ceeaed6db3b4aca000fa2f53ba37764dedd204e1fd27e"><div class="ttname"><a href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4aca000fa2f53ba37764dedd204e1fd27e">vg::Genotyper::Representative</a></div><div class="ttdeci">@ Representative</div><div class="ttdef"><b>Definition:</b> genotyper.hpp:140</div></div>
</div><!-- fragment -->
</div>
</div>
<a id="ac177e2664adf7be9b92939148ea77c65"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac177e2664adf7be9b92939148ea77c65">&#9670;&nbsp;</a></span>all_snarls</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unordered_set&lt;const <a class="el" href="structvg_1_1Snarl.html">Snarl</a>*&gt; vg::Genotyper::all_snarls</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a384f22f52c2a1f78b7ca613468e2e186"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a384f22f52c2a1f78b7ca613468e2e186">&#9670;&nbsp;</a></span>dagify_steps</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vg::Genotyper::dagify_steps = 1</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a3c40092b41694cacc7e531bf7f9fb8aa"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3c40092b41694cacc7e531bf7f9fb8aa">&#9670;&nbsp;</a></span>default_sequence_quality</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vg::Genotyper::default_sequence_quality = 15</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a853fb8840552901b359153ec80aa81e1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a853fb8840552901b359153ec80aa81e1">&#9670;&nbsp;</a></span>deleted_prior_logprob</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vg::Genotyper::deleted_prior_logprob = <a class="el" href="namespacevg.html#a476a2102f4f5aa82c06be87780ae6d42">prob_to_logprob</a>(0.1)</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="ac748204c3fb19573569172ee57b1d082"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac748204c3fb19573569172ee57b1d082">&#9670;&nbsp;</a></span>diploid_prior_logprob</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vg::Genotyper::diploid_prior_logprob = <a class="el" href="namespacevg.html#a476a2102f4f5aa82c06be87780ae6d42">prob_to_logprob</a>(0.999)</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a76e36b8cc1f5a8a5522c5ae3f57b5464"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a76e36b8cc1f5a8a5522c5ae3f57b5464">&#9670;&nbsp;</a></span>haploid_prior_logprob</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vg::Genotyper::haploid_prior_logprob = <a class="el" href="namespacevg.html#a476a2102f4f5aa82c06be87780ae6d42">prob_to_logprob</a>(0.5)</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a21799af8d6da86a3bbcfffb1060d27cf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a21799af8d6da86a3bbcfffb1060d27cf">&#9670;&nbsp;</a></span>het_prior_logprob</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vg::Genotyper::het_prior_logprob = <a class="el" href="namespacevg.html#a476a2102f4f5aa82c06be87780ae6d42">prob_to_logprob</a>(0.1)</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a96285d96b960d50199ad3167c20ddf91"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a96285d96b960d50199ad3167c20ddf91">&#9670;&nbsp;</a></span>max_het_bias</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vg::Genotyper::max_het_bias = 4</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a500af110546f9572fe40d5ef098552d9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a500af110546f9572fe40d5ef098552d9">&#9670;&nbsp;</a></span>max_path_search_steps</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t vg::Genotyper::max_path_search_steps = 100</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="afffa8d75fd4155bc7d420a41d92f198c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afffa8d75fd4155bc7d420a41d92f198c">&#9670;&nbsp;</a></span>min_recurrence</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vg::Genotyper::min_recurrence = 2</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a8fc1fdccfa9829bb29903ed5d7b21b28"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8fc1fdccfa9829bb29903ed5d7b21b28">&#9670;&nbsp;</a></span>min_score_per_base</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vg::Genotyper::min_score_per_base = 0.90</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a6b1feebb10944e684e417115e880fe39"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6b1feebb10944e684e417115e880fe39">&#9670;&nbsp;</a></span>min_unique_per_strand</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vg::Genotyper::min_unique_per_strand = 2</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a91d6174a20ca4f121f68766fbf7fb4b0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a91d6174a20ca4f121f68766fbf7fb4b0">&#9670;&nbsp;</a></span>normal_aligners</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt;<a class="el" href="classvg_1_1Aligner.html">Aligner</a>&gt; vg::Genotyper::normal_aligners</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a97aef0fe84f0514e17e2ec8e00162225"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a97aef0fe84f0514e17e2ec8e00162225">&#9670;&nbsp;</a></span>polyploid_prior_success_logprob</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vg::Genotyper::polyploid_prior_success_logprob = <a class="el" href="namespacevg.html#a476a2102f4f5aa82c06be87780ae6d42">prob_to_logprob</a>(0.5)</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a2705913bdfb5f6679afb5475a22ecf61"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2705913bdfb5f6679afb5475a22ecf61">&#9670;&nbsp;</a></span>quality_aligners</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt;<a class="el" href="classvg_1_1QualAdjAligner.html">QualAdjAligner</a>&gt; vg::Genotyper::quality_aligners</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a79ac39be882405a426afaeeef65c9660"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a79ac39be882405a426afaeeef65c9660">&#9670;&nbsp;</a></span>realign_indels</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Genotyper::realign_indels = false</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a72599c00acaed553e0d235e5c73d4b82"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a72599c00acaed553e0d235e5c73d4b82">&#9670;&nbsp;</a></span>show_progress</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Genotyper::show_progress = false</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="ac0d4917131823fc2dab43927347d4623"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac0d4917131823fc2dab43927347d4623">&#9670;&nbsp;</a></span>snarl_reference_length_histogram</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unordered_map&lt;size_t, size_t&gt; vg::Genotyper::snarl_reference_length_histogram</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a7ec1915674299d727c61377e91385974"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7ec1915674299d727c61377e91385974">&#9670;&nbsp;</a></span>snarl_traversals</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unordered_set&lt;const <a class="el" href="structvg_1_1Snarl.html">Snarl</a>*&gt; vg::Genotyper::snarl_traversals</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="ad14928554aace83524e80d8c109e17b2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad14928554aace83524e80d8c109e17b2">&#9670;&nbsp;</a></span>translator</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classvg_1_1Translator.html">Translator</a> vg::Genotyper::translator</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a5b040c1636f479b52dbdb641e82f74dd"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5b040c1636f479b52dbdb641e82f74dd">&#9670;&nbsp;</a></span>traversal_alg</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classvg_1_1Genotyper.html#a6e4dce6cd9cc24ab6857ceeaed6db3b4">TraversalAlg</a> vg::Genotyper::traversal_alg = TraversalAlg::Reads</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="ab3e4270419b361c9455125ca305709e2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab3e4270419b361c9455125ca305709e2">&#9670;&nbsp;</a></span>unfold_max_length</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vg::Genotyper::unfold_max_length = 200</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a5991ce66cc2a9ea45bd27d288dd7d9b8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5991ce66cc2a9ea45bd27d288dd7d9b8">&#9670;&nbsp;</a></span>use_mapq</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Genotyper::use_mapq = false</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>src/<a class="el" href="genotyper_8hpp.html">genotyper.hpp</a></li>
<li>src/<a class="el" href="genotyper_8cpp.html">genotyper.cpp</a></li>
</ul>
</div><!-- contents -->
<!-- HTML footer for doxygen 1.8.9.1-->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.9.1
</small></address>
</body>
</html>
