<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>vg: vg::Alignment Struct Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">vg
   </div>
   <div id="projectbrief">tools for working with variation graphs</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacevg.html">vg</a></li><li class="navelem"><a class="el" href="structvg_1_1Alignment.html">Alignment</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="structvg_1_1Alignment-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">vg::Alignment Struct Reference</div>  </div>
</div><!--header-->
<div class="contents">
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:a2ae1f06d48c37997cee72b7cd3210a83"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a2ae1f06d48c37997cee72b7cd3210a83">sequence</a> = 1</td></tr>
<tr class="memdesc:a2ae1f06d48c37997cee72b7cd3210a83"><td class="mdescLeft">&#160;</td><td class="mdescRight">The sequence that has been aligned.  <a href="structvg_1_1Alignment.html#a2ae1f06d48c37997cee72b7cd3210a83">More...</a><br /></td></tr>
<tr class="separator:a2ae1f06d48c37997cee72b7cd3210a83"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4736db5fe121ef57784c18c3c27aea79"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structvg_1_1Path.html">Path</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a4736db5fe121ef57784c18c3c27aea79">path</a> = 2</td></tr>
<tr class="memdesc:a4736db5fe121ef57784c18c3c27aea79"><td class="mdescLeft">&#160;</td><td class="mdescRight">The <a class="el" href="structvg_1_1Path.html">Path</a> that the sequence follows in the graph it has been aligned to, containing the <code><a class="el" href="structvg_1_1Edit.html">Edit</a></code>s that modify the graph to produce the sequence.  <a href="structvg_1_1Alignment.html#a4736db5fe121ef57784c18c3c27aea79">More...</a><br /></td></tr>
<tr class="separator:a4736db5fe121ef57784c18c3c27aea79"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad20cac38e0bd25d2b32862fe1f5409ab"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#ad20cac38e0bd25d2b32862fe1f5409ab">name</a> = 3</td></tr>
<tr class="memdesc:ad20cac38e0bd25d2b32862fe1f5409ab"><td class="mdescLeft">&#160;</td><td class="mdescRight">The name of the sequence that has been aligned. Similar to read name in BAM.  <a href="structvg_1_1Alignment.html#ad20cac38e0bd25d2b32862fe1f5409ab">More...</a><br /></td></tr>
<tr class="separator:ad20cac38e0bd25d2b32862fe1f5409ab"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afe4dfd54e11b8139c3ccbb493faf98dc"><td class="memItemLeft" align="right" valign="top">bytes&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#afe4dfd54e11b8139c3ccbb493faf98dc">quality</a> = 4</td></tr>
<tr class="memdesc:afe4dfd54e11b8139c3ccbb493faf98dc"><td class="mdescLeft">&#160;</td><td class="mdescRight">The quality scores for the sequence, as values on a 0-255 scale.  <a href="structvg_1_1Alignment.html#afe4dfd54e11b8139c3ccbb493faf98dc">More...</a><br /></td></tr>
<tr class="separator:afe4dfd54e11b8139c3ccbb493faf98dc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a78851a70cc8a3e8f3e71409693130230"><td class="memItemLeft" align="right" valign="top">int32&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a78851a70cc8a3e8f3e71409693130230">mapping_quality</a> = 5</td></tr>
<tr class="memdesc:a78851a70cc8a3e8f3e71409693130230"><td class="mdescLeft">&#160;</td><td class="mdescRight">The mapping quality score for the alignment, in Phreds.  <a href="structvg_1_1Alignment.html#a78851a70cc8a3e8f3e71409693130230">More...</a><br /></td></tr>
<tr class="separator:a78851a70cc8a3e8f3e71409693130230"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5c57b95a7ea122139a092427b592ebeb"><td class="memItemLeft" align="right" valign="top">int32&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a5c57b95a7ea122139a092427b592ebeb">score</a> = 6</td></tr>
<tr class="memdesc:a5c57b95a7ea122139a092427b592ebeb"><td class="mdescLeft">&#160;</td><td class="mdescRight">The score for the alignment, in points.  <a href="structvg_1_1Alignment.html#a5c57b95a7ea122139a092427b592ebeb">More...</a><br /></td></tr>
<tr class="separator:a5c57b95a7ea122139a092427b592ebeb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a17a81788105d3cb6a777ed6a202cd36f"><td class="memItemLeft" align="right" valign="top">int32&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a17a81788105d3cb6a777ed6a202cd36f">query_position</a> = 7</td></tr>
<tr class="memdesc:a17a81788105d3cb6a777ed6a202cd36f"><td class="mdescLeft">&#160;</td><td class="mdescRight">The offset in the query at which this <a class="el" href="structvg_1_1Alignment.html">Alignment</a> occurs.  <a href="structvg_1_1Alignment.html#a17a81788105d3cb6a777ed6a202cd36f">More...</a><br /></td></tr>
<tr class="separator:a17a81788105d3cb6a777ed6a202cd36f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8374111b6348509e8e9ecca0cd1ac046"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a8374111b6348509e8e9ecca0cd1ac046">reserved</a></td></tr>
<tr class="memdesc:a8374111b6348509e8e9ecca0cd1ac046"><td class="mdescLeft">&#160;</td><td class="mdescRight">Old field 8 has been removed.  <a href="structvg_1_1Alignment.html#a8374111b6348509e8e9ecca0cd1ac046">More...</a><br /></td></tr>
<tr class="separator:a8374111b6348509e8e9ecca0cd1ac046"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad823d9ed3f413faaf35700a41cfab6e7"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#ad823d9ed3f413faaf35700a41cfab6e7">sample_name</a> = 9</td></tr>
<tr class="memdesc:ad823d9ed3f413faaf35700a41cfab6e7"><td class="mdescLeft">&#160;</td><td class="mdescRight">The name of the sample that produced the aligned read.  <a href="structvg_1_1Alignment.html#ad823d9ed3f413faaf35700a41cfab6e7">More...</a><br /></td></tr>
<tr class="separator:ad823d9ed3f413faaf35700a41cfab6e7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a95102c70d75046f86a2f0a829ba1efe0"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a95102c70d75046f86a2f0a829ba1efe0">read_group</a> = 10</td></tr>
<tr class="memdesc:a95102c70d75046f86a2f0a829ba1efe0"><td class="mdescLeft">&#160;</td><td class="mdescRight">The name of the read group to which the aligned read belongs.  <a href="structvg_1_1Alignment.html#a95102c70d75046f86a2f0a829ba1efe0">More...</a><br /></td></tr>
<tr class="separator:a95102c70d75046f86a2f0a829ba1efe0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a82a59f4fec8b7fcd069563d99fa04228"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structvg_1_1Alignment.html">Alignment</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a82a59f4fec8b7fcd069563d99fa04228">fragment_prev</a> = 11</td></tr>
<tr class="memdesc:a82a59f4fec8b7fcd069563d99fa04228"><td class="mdescLeft">&#160;</td><td class="mdescRight">The previous <a class="el" href="structvg_1_1Alignment.html">Alignment</a> in the fragment. Contains just enough information to locate the full <a class="el" href="structvg_1_1Alignment.html">Alignment</a>; e.g. contains an <a class="el" href="structvg_1_1Alignment.html">Alignment</a> with only a name, or only a graph mapping position.  <a href="structvg_1_1Alignment.html#a82a59f4fec8b7fcd069563d99fa04228">More...</a><br /></td></tr>
<tr class="separator:a82a59f4fec8b7fcd069563d99fa04228"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a79ace3e74254f2a6e1ed5bdf09eb226f"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structvg_1_1Alignment.html">Alignment</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a79ace3e74254f2a6e1ed5bdf09eb226f">fragment_next</a> = 12</td></tr>
<tr class="memdesc:a79ace3e74254f2a6e1ed5bdf09eb226f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Similarly, the next <a class="el" href="structvg_1_1Alignment.html">Alignment</a> in the fragment.  <a href="structvg_1_1Alignment.html#a79ace3e74254f2a6e1ed5bdf09eb226f">More...</a><br /></td></tr>
<tr class="separator:a79ace3e74254f2a6e1ed5bdf09eb226f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a59a4b8c7367fb468e46bcf6ae292888f"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a59a4b8c7367fb468e46bcf6ae292888f">is_secondary</a> = 15</td></tr>
<tr class="memdesc:a59a4b8c7367fb468e46bcf6ae292888f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flag marking the <a class="el" href="structvg_1_1Alignment.html">Alignment</a> as secondary. All but one maximal-scoring alignment of a given read in a GAM file must be secondary.  <a href="structvg_1_1Alignment.html#a59a4b8c7367fb468e46bcf6ae292888f">More...</a><br /></td></tr>
<tr class="separator:a59a4b8c7367fb468e46bcf6ae292888f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a06341663199b9893a9f2b288a768cfe3"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a06341663199b9893a9f2b288a768cfe3">identity</a> = 16</td></tr>
<tr class="memdesc:a06341663199b9893a9f2b288a768cfe3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Portion of aligned bases that are perfect matches, or 0 if no bases are aligned.  <a href="structvg_1_1Alignment.html#a06341663199b9893a9f2b288a768cfe3">More...</a><br /></td></tr>
<tr class="separator:a06341663199b9893a9f2b288a768cfe3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a600fa84d6905316c0964dd035ac48687"><td class="memItemLeft" align="right" valign="top">repeated&lt; <a class="el" href="structvg_1_1Path.html">Path</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a600fa84d6905316c0964dd035ac48687">fragment</a> = 17</td></tr>
<tr class="memdesc:a600fa84d6905316c0964dd035ac48687"><td class="mdescLeft">&#160;</td><td class="mdescRight">An estimate of the length of the fragment, if this <a class="el" href="structvg_1_1Alignment.html">Alignment</a> is paired.  <a href="structvg_1_1Alignment.html#a600fa84d6905316c0964dd035ac48687">More...</a><br /></td></tr>
<tr class="separator:a600fa84d6905316c0964dd035ac48687"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afe47b70cd7b78e97a9cf18e9c60ef4f0"><td class="memItemLeft" align="right" valign="top">repeated&lt; <a class="el" href="structvg_1_1Locus.html">Locus</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#afe47b70cd7b78e97a9cf18e9c60ef4f0">locus</a> = 18</td></tr>
<tr class="memdesc:afe47b70cd7b78e97a9cf18e9c60ef4f0"><td class="mdescLeft">&#160;</td><td class="mdescRight">The loci that this alignment supports. TODO: get rid of this, we have annotations in our data model again.  <a href="structvg_1_1Alignment.html#afe47b70cd7b78e97a9cf18e9c60ef4f0">More...</a><br /></td></tr>
<tr class="separator:afe47b70cd7b78e97a9cf18e9c60ef4f0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a584c3aa15a5c1386d14a97c0ce85c0cf"><td class="memItemLeft" align="right" valign="top">repeated&lt; <a class="el" href="structvg_1_1Position.html">Position</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a584c3aa15a5c1386d14a97c0ce85c0cf">refpos</a> = 19</td></tr>
<tr class="memdesc:a584c3aa15a5c1386d14a97c0ce85c0cf"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="structvg_1_1Position.html">Position</a> of the alignment in reference paths embedded in graph. Each position has a path name, and the <a class="el" href="structvg_1_1Alignment.html">Alignment</a>'s minimum position along the path as an offset.  <a href="structvg_1_1Alignment.html#a584c3aa15a5c1386d14a97c0ce85c0cf">More...</a><br /></td></tr>
<tr class="separator:a584c3aa15a5c1386d14a97c0ce85c0cf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4b8f47ac6136d64e1a1d8cd187b534d6"><td class="memItemLeft" align="right" valign="top">repeated&lt; <a class="el" href="structvg_1_1Alignment.html">Alignment</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a4b8f47ac6136d64e1a1d8cd187b534d6">supplementary</a> = 38</td></tr>
<tr class="memdesc:a4b8f47ac6136d64e1a1d8cd187b534d6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Discontiguous alignment(s) that form a chimeric alignment of the sequence along with the containing <a class="el" href="structvg_1_1Alignment.html">Alignment</a> itself.  <a href="structvg_1_1Alignment.html#a4b8f47ac6136d64e1a1d8cd187b534d6">More...</a><br /></td></tr>
<tr class="separator:a4b8f47ac6136d64e1a1d8cd187b534d6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2bfa3fffb5f4ea6ab3e98a6340ad16bb"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a2bfa3fffb5f4ea6ab3e98a6340ad16bb">read_paired</a> = 20</td></tr>
<tr class="memdesc:a2bfa3fffb5f4ea6ab3e98a6340ad16bb"><td class="mdescLeft">&#160;</td><td class="mdescRight">SAMTools-style flags.  <a href="structvg_1_1Alignment.html#a2bfa3fffb5f4ea6ab3e98a6340ad16bb">More...</a><br /></td></tr>
<tr class="separator:a2bfa3fffb5f4ea6ab3e98a6340ad16bb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af2fc72c769b77c355632e6aab7573046"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#af2fc72c769b77c355632e6aab7573046">read_mapped</a> = 21</td></tr>
<tr class="separator:af2fc72c769b77c355632e6aab7573046"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a061bf4ec65ccf7ff88e73e9e4ff268b7"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a061bf4ec65ccf7ff88e73e9e4ff268b7">mate_unmapped</a> = 22</td></tr>
<tr class="separator:a061bf4ec65ccf7ff88e73e9e4ff268b7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a10b4d599e1812ee007e348f8ce8bfe49"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a10b4d599e1812ee007e348f8ce8bfe49">read_on_reverse_strand</a> = 23</td></tr>
<tr class="separator:a10b4d599e1812ee007e348f8ce8bfe49"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aef312ed490cb8adf8b89eb36b46ee7d6"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#aef312ed490cb8adf8b89eb36b46ee7d6">mate_on_reverse_strand</a> = 24</td></tr>
<tr class="separator:aef312ed490cb8adf8b89eb36b46ee7d6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6cb12eae3f4fa017bc98b59a91a5414f"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a6cb12eae3f4fa017bc98b59a91a5414f">soft_clipped</a> = 25</td></tr>
<tr class="separator:a6cb12eae3f4fa017bc98b59a91a5414f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5cf554f770c01df31d8b4046b3ed5f50"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a5cf554f770c01df31d8b4046b3ed5f50">discordant_insert_size</a> = 26</td></tr>
<tr class="separator:a5cf554f770c01df31d8b4046b3ed5f50"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2d744ce39b375cdeb8adfc3721a18eb0"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a2d744ce39b375cdeb8adfc3721a18eb0">uniqueness</a> = 27</td></tr>
<tr class="memdesc:a2d744ce39b375cdeb8adfc3721a18eb0"><td class="mdescLeft">&#160;</td><td class="mdescRight">The fraction of bases in the alignment that are covered by MEMs with &lt;=1 total hits in the graph.  <a href="structvg_1_1Alignment.html#a2d744ce39b375cdeb8adfc3721a18eb0">More...</a><br /></td></tr>
<tr class="separator:a2d744ce39b375cdeb8adfc3721a18eb0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a07751b50cfb9c985fccc2c701bfc55fc"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a07751b50cfb9c985fccc2c701bfc55fc">correct</a> = 28</td></tr>
<tr class="memdesc:a07751b50cfb9c985fccc2c701bfc55fc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Correctness metric 1 = perfectly aligned to truth, 0 = not overlapping true alignment.  <a href="structvg_1_1Alignment.html#a07751b50cfb9c985fccc2c701bfc55fc">More...</a><br /></td></tr>
<tr class="separator:a07751b50cfb9c985fccc2c701bfc55fc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a58502a21a2588a5ac8baaa6e31c81eb4"><td class="memItemLeft" align="right" valign="top">repeated&lt; int32 &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a58502a21a2588a5ac8baaa6e31c81eb4">secondary_score</a> = 29</td></tr>
<tr class="memdesc:a58502a21a2588a5ac8baaa6e31c81eb4"><td class="mdescLeft">&#160;</td><td class="mdescRight">The ordered list of scores of secondary mappings.  <a href="structvg_1_1Alignment.html#a58502a21a2588a5ac8baaa6e31c81eb4">More...</a><br /></td></tr>
<tr class="separator:a58502a21a2588a5ac8baaa6e31c81eb4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2a105371920e65f6764f8b723ae6648d"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a2a105371920e65f6764f8b723ae6648d">fragment_score</a> = 30</td></tr>
<tr class="memdesc:a2a105371920e65f6764f8b723ae6648d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Score under the given fragment model, assume higher is better.  <a href="structvg_1_1Alignment.html#a2a105371920e65f6764f8b723ae6648d">More...</a><br /></td></tr>
<tr class="separator:a2a105371920e65f6764f8b723ae6648d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a95ed4d8c7a57e34a96871b74ff851821"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a95ed4d8c7a57e34a96871b74ff851821">mate_mapped_to_disjoint_subgraph</a> = 31</td></tr>
<tr class="separator:a95ed4d8c7a57e34a96871b74ff851821"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab13fb82c1707e0c6934d5c5ffd4c3ffa"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#ab13fb82c1707e0c6934d5c5ffd4c3ffa">fragment_length_distribution</a> = 32</td></tr>
<tr class="memdesc:ab13fb82c1707e0c6934d5c5ffd4c3ffa"><td class="mdescLeft">&#160;</td><td class="mdescRight">The fragment length distribution under which a paired-end alignment was aligned.  <a href="structvg_1_1Alignment.html#ab13fb82c1707e0c6934d5c5ffd4c3ffa">More...</a><br /></td></tr>
<tr class="separator:ab13fb82c1707e0c6934d5c5ffd4c3ffa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9db86772c1bd48808c0588f96f3c0865"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structvg_1_1Alignment.html#a8374111b6348509e8e9ecca0cd1ac046">reserved</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a9db86772c1bd48808c0588f96f3c0865">to</a></td></tr>
<tr class="memdesc:a9db86772c1bd48808c0588f96f3c0865"><td class="mdescLeft">&#160;</td><td class="mdescRight">Haplotype-scoring-related fields migrated to the annotation system.  <a href="structvg_1_1Alignment.html#a9db86772c1bd48808c0588f96f3c0865">More...</a><br /></td></tr>
<tr class="separator:a9db86772c1bd48808c0588f96f3c0865"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0c8b04a220715a894a19dcc70ed581c0"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a0c8b04a220715a894a19dcc70ed581c0">time_used</a> = 35</td></tr>
<tr class="memdesc:a0c8b04a220715a894a19dcc70ed581c0"><td class="mdescLeft">&#160;</td><td class="mdescRight">The time this alignment took.  <a href="structvg_1_1Alignment.html#a0c8b04a220715a894a19dcc70ed581c0">More...</a><br /></td></tr>
<tr class="separator:a0c8b04a220715a894a19dcc70ed581c0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8957ee4a5a3ef66c93f6b6b685774df2"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structvg_1_1Position.html">Position</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a8957ee4a5a3ef66c93f6b6b685774df2">to_correct</a> = 36</td></tr>
<tr class="memdesc:a8957ee4a5a3ef66c93f6b6b685774df2"><td class="mdescLeft">&#160;</td><td class="mdescRight">A path/offset/orientation pair specifying the distance to the correct alignment.  <a href="structvg_1_1Alignment.html#a8957ee4a5a3ef66c93f6b6b685774df2">More...</a><br /></td></tr>
<tr class="separator:a8957ee4a5a3ef66c93f6b6b685774df2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9ec303e39d844ec1e6289ddfbcdd7911"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#a9ec303e39d844ec1e6289ddfbcdd7911">correctly_mapped</a> = 37</td></tr>
<tr class="memdesc:a9ec303e39d844ec1e6289ddfbcdd7911"><td class="mdescLeft">&#160;</td><td class="mdescRight">This can be set to true to annotate the <a class="el" href="structvg_1_1Alignment.html">Alignment</a> as having been mapped correctly.  <a href="structvg_1_1Alignment.html#a9ec303e39d844ec1e6289ddfbcdd7911">More...</a><br /></td></tr>
<tr class="separator:a9ec303e39d844ec1e6289ddfbcdd7911"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af531ff2b8c2b08c78a8b3095fd52c0d8"><td class="memItemLeft" align="right" valign="top">google protobuf Struct&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1Alignment.html#af531ff2b8c2b08c78a8b3095fd52c0d8">annotation</a> = 100</td></tr>
<tr class="memdesc:af531ff2b8c2b08c78a8b3095fd52c0d8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Annotations carried along with the <a class="el" href="structvg_1_1Alignment.html">Alignment</a>.  <a href="structvg_1_1Alignment.html#af531ff2b8c2b08c78a8b3095fd52c0d8">More...</a><br /></td></tr>
<tr class="separator:af531ff2b8c2b08c78a8b3095fd52c0d8"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Alignments link query strings, such as other genomes or reads, to <a class="el" href="classvg_1_1Paths.html">Paths</a>. </p>
</div><h2 class="groupheader">Member Data Documentation</h2>
<a id="af531ff2b8c2b08c78a8b3095fd52c0d8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af531ff2b8c2b08c78a8b3095fd52c0d8">&#9670;&nbsp;</a></span>annotation</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">google protobuf Struct vg::Alignment::annotation = 100</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Annotations carried along with the <a class="el" href="structvg_1_1Alignment.html">Alignment</a>. </p>

</div>
</div>
<a id="a07751b50cfb9c985fccc2c701bfc55fc"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a07751b50cfb9c985fccc2c701bfc55fc">&#9670;&nbsp;</a></span>correct</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vg::Alignment::correct = 28</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Correctness metric 1 = perfectly aligned to truth, 0 = not overlapping true alignment. </p>

</div>
</div>
<a id="a9ec303e39d844ec1e6289ddfbcdd7911"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9ec303e39d844ec1e6289ddfbcdd7911">&#9670;&nbsp;</a></span>correctly_mapped</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Alignment::correctly_mapped = 37</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>This can be set to true to annotate the <a class="el" href="structvg_1_1Alignment.html">Alignment</a> as having been mapped correctly. </p>

</div>
</div>
<a id="a5cf554f770c01df31d8b4046b3ed5f50"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5cf554f770c01df31d8b4046b3ed5f50">&#9670;&nbsp;</a></span>discordant_insert_size</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Alignment::discordant_insert_size = 26</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a600fa84d6905316c0964dd035ac48687"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a600fa84d6905316c0964dd035ac48687">&#9670;&nbsp;</a></span>fragment</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">repeated&lt;<a class="el" href="structvg_1_1Path.html">Path</a>&gt; vg::Alignment::fragment = 17</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>An estimate of the length of the fragment, if this <a class="el" href="structvg_1_1Alignment.html">Alignment</a> is paired. </p>

</div>
</div>
<a id="ab13fb82c1707e0c6934d5c5ffd4c3ffa"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab13fb82c1707e0c6934d5c5ffd4c3ffa">&#9670;&nbsp;</a></span>fragment_length_distribution</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">string vg::Alignment::fragment_length_distribution = 32</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The fragment length distribution under which a paired-end alignment was aligned. </p>

</div>
</div>
<a id="a79ace3e74254f2a6e1ed5bdf09eb226f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a79ace3e74254f2a6e1ed5bdf09eb226f">&#9670;&nbsp;</a></span>fragment_next</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structvg_1_1Alignment.html">Alignment</a> vg::Alignment::fragment_next = 12</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Similarly, the next <a class="el" href="structvg_1_1Alignment.html">Alignment</a> in the fragment. </p>

</div>
</div>
<a id="a82a59f4fec8b7fcd069563d99fa04228"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a82a59f4fec8b7fcd069563d99fa04228">&#9670;&nbsp;</a></span>fragment_prev</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structvg_1_1Alignment.html">Alignment</a> vg::Alignment::fragment_prev = 11</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The previous <a class="el" href="structvg_1_1Alignment.html">Alignment</a> in the fragment. Contains just enough information to locate the full <a class="el" href="structvg_1_1Alignment.html">Alignment</a>; e.g. contains an <a class="el" href="structvg_1_1Alignment.html">Alignment</a> with only a name, or only a graph mapping position. </p>

</div>
</div>
<a id="a2a105371920e65f6764f8b723ae6648d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2a105371920e65f6764f8b723ae6648d">&#9670;&nbsp;</a></span>fragment_score</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vg::Alignment::fragment_score = 30</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Score under the given fragment model, assume higher is better. </p>

</div>
</div>
<a id="a06341663199b9893a9f2b288a768cfe3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a06341663199b9893a9f2b288a768cfe3">&#9670;&nbsp;</a></span>identity</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vg::Alignment::identity = 16</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Portion of aligned bases that are perfect matches, or 0 if no bases are aligned. </p>

</div>
</div>
<a id="a59a4b8c7367fb468e46bcf6ae292888f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a59a4b8c7367fb468e46bcf6ae292888f">&#9670;&nbsp;</a></span>is_secondary</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Alignment::is_secondary = 15</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Flag marking the <a class="el" href="structvg_1_1Alignment.html">Alignment</a> as secondary. All but one maximal-scoring alignment of a given read in a GAM file must be secondary. </p>

</div>
</div>
<a id="afe47b70cd7b78e97a9cf18e9c60ef4f0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afe47b70cd7b78e97a9cf18e9c60ef4f0">&#9670;&nbsp;</a></span>locus</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">repeated&lt;<a class="el" href="structvg_1_1Locus.html">Locus</a>&gt; vg::Alignment::locus = 18</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The loci that this alignment supports. TODO: get rid of this, we have annotations in our data model again. </p>

</div>
</div>
<a id="a78851a70cc8a3e8f3e71409693130230"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a78851a70cc8a3e8f3e71409693130230">&#9670;&nbsp;</a></span>mapping_quality</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int32 vg::Alignment::mapping_quality = 5</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The mapping quality score for the alignment, in Phreds. </p>

</div>
</div>
<a id="a95ed4d8c7a57e34a96871b74ff851821"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a95ed4d8c7a57e34a96871b74ff851821">&#9670;&nbsp;</a></span>mate_mapped_to_disjoint_subgraph</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Alignment::mate_mapped_to_disjoint_subgraph = 31</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="aef312ed490cb8adf8b89eb36b46ee7d6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aef312ed490cb8adf8b89eb36b46ee7d6">&#9670;&nbsp;</a></span>mate_on_reverse_strand</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Alignment::mate_on_reverse_strand = 24</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a061bf4ec65ccf7ff88e73e9e4ff268b7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a061bf4ec65ccf7ff88e73e9e4ff268b7">&#9670;&nbsp;</a></span>mate_unmapped</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Alignment::mate_unmapped = 22</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="ad20cac38e0bd25d2b32862fe1f5409ab"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad20cac38e0bd25d2b32862fe1f5409ab">&#9670;&nbsp;</a></span>name</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">string vg::Alignment::name = 3</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The name of the sequence that has been aligned. Similar to read name in BAM. </p>

</div>
</div>
<a id="a4736db5fe121ef57784c18c3c27aea79"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4736db5fe121ef57784c18c3c27aea79">&#9670;&nbsp;</a></span>path</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structvg_1_1Path.html">Path</a> vg::Alignment::path = 2</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The <a class="el" href="structvg_1_1Path.html">Path</a> that the sequence follows in the graph it has been aligned to, containing the <code><a class="el" href="structvg_1_1Edit.html">Edit</a></code>s that modify the graph to produce the sequence. </p>

</div>
</div>
<a id="afe4dfd54e11b8139c3ccbb493faf98dc"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afe4dfd54e11b8139c3ccbb493faf98dc">&#9670;&nbsp;</a></span>quality</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bytes vg::Alignment::quality = 4</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The quality scores for the sequence, as values on a 0-255 scale. </p>

</div>
</div>
<a id="a17a81788105d3cb6a777ed6a202cd36f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a17a81788105d3cb6a777ed6a202cd36f">&#9670;&nbsp;</a></span>query_position</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int32 vg::Alignment::query_position = 7</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The offset in the query at which this <a class="el" href="structvg_1_1Alignment.html">Alignment</a> occurs. </p>

</div>
</div>
<a id="a95102c70d75046f86a2f0a829ba1efe0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a95102c70d75046f86a2f0a829ba1efe0">&#9670;&nbsp;</a></span>read_group</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">string vg::Alignment::read_group = 10</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The name of the read group to which the aligned read belongs. </p>

</div>
</div>
<a id="af2fc72c769b77c355632e6aab7573046"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af2fc72c769b77c355632e6aab7573046">&#9670;&nbsp;</a></span>read_mapped</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Alignment::read_mapped = 21</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a10b4d599e1812ee007e348f8ce8bfe49"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a10b4d599e1812ee007e348f8ce8bfe49">&#9670;&nbsp;</a></span>read_on_reverse_strand</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Alignment::read_on_reverse_strand = 23</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a2bfa3fffb5f4ea6ab3e98a6340ad16bb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2bfa3fffb5f4ea6ab3e98a6340ad16bb">&#9670;&nbsp;</a></span>read_paired</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Alignment::read_paired = 20</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>SAMTools-style flags. </p>

</div>
</div>
<a id="a584c3aa15a5c1386d14a97c0ce85c0cf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a584c3aa15a5c1386d14a97c0ce85c0cf">&#9670;&nbsp;</a></span>refpos</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">repeated&lt;<a class="el" href="structvg_1_1Position.html">Position</a>&gt; vg::Alignment::refpos = 19</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><a class="el" href="structvg_1_1Position.html">Position</a> of the alignment in reference paths embedded in graph. Each position has a path name, and the <a class="el" href="structvg_1_1Alignment.html">Alignment</a>'s minimum position along the path as an offset. </p>

</div>
</div>
<a id="a8374111b6348509e8e9ecca0cd1ac046"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8374111b6348509e8e9ecca0cd1ac046">&#9670;&nbsp;</a></span>reserved</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vg::Alignment::reserved</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Old field 8 has been removed. </p>

</div>
</div>
<a id="ad823d9ed3f413faaf35700a41cfab6e7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad823d9ed3f413faaf35700a41cfab6e7">&#9670;&nbsp;</a></span>sample_name</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">string vg::Alignment::sample_name = 9</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The name of the sample that produced the aligned read. </p>

</div>
</div>
<a id="a5c57b95a7ea122139a092427b592ebeb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5c57b95a7ea122139a092427b592ebeb">&#9670;&nbsp;</a></span>score</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int32 vg::Alignment::score = 6</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The score for the alignment, in points. </p>

</div>
</div>
<a id="a58502a21a2588a5ac8baaa6e31c81eb4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a58502a21a2588a5ac8baaa6e31c81eb4">&#9670;&nbsp;</a></span>secondary_score</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">repeated&lt;int32&gt; vg::Alignment::secondary_score = 29</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The ordered list of scores of secondary mappings. </p>

</div>
</div>
<a id="a2ae1f06d48c37997cee72b7cd3210a83"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2ae1f06d48c37997cee72b7cd3210a83">&#9670;&nbsp;</a></span>sequence</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">string vg::Alignment::sequence = 1</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The sequence that has been aligned. </p>

</div>
</div>
<a id="a6cb12eae3f4fa017bc98b59a91a5414f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6cb12eae3f4fa017bc98b59a91a5414f">&#9670;&nbsp;</a></span>soft_clipped</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Alignment::soft_clipped = 25</td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a id="a4b8f47ac6136d64e1a1d8cd187b534d6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4b8f47ac6136d64e1a1d8cd187b534d6">&#9670;&nbsp;</a></span>supplementary</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">repeated&lt;<a class="el" href="structvg_1_1Alignment.html">Alignment</a>&gt; vg::Alignment::supplementary = 38</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Discontiguous alignment(s) that form a chimeric alignment of the sequence along with the containing <a class="el" href="structvg_1_1Alignment.html">Alignment</a> itself. </p>

</div>
</div>
<a id="a0c8b04a220715a894a19dcc70ed581c0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0c8b04a220715a894a19dcc70ed581c0">&#9670;&nbsp;</a></span>time_used</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vg::Alignment::time_used = 35</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The time this alignment took. </p>

</div>
</div>
<a id="a9db86772c1bd48808c0588f96f3c0865"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9db86772c1bd48808c0588f96f3c0865">&#9670;&nbsp;</a></span>to</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structvg_1_1Alignment.html#a8374111b6348509e8e9ecca0cd1ac046">reserved</a> vg::Alignment::to</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Haplotype-scoring-related fields migrated to the annotation system. </p>

</div>
</div>
<a id="a8957ee4a5a3ef66c93f6b6b685774df2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8957ee4a5a3ef66c93f6b6b685774df2">&#9670;&nbsp;</a></span>to_correct</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structvg_1_1Position.html">Position</a> vg::Alignment::to_correct = 36</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>A path/offset/orientation pair specifying the distance to the correct alignment. </p>

</div>
</div>
<a id="a2d744ce39b375cdeb8adfc3721a18eb0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2d744ce39b375cdeb8adfc3721a18eb0">&#9670;&nbsp;</a></span>uniqueness</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vg::Alignment::uniqueness = 27</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>The fraction of bases in the alignment that are covered by MEMs with &lt;=1 total hits in the graph. </p>

</div>
</div>
<hr/>The documentation for this struct was generated from the following file:<ul>
<li>deps/libvgio/deps/<a class="el" href="vg_8proto.html">vg.proto</a></li>
</ul>
</div><!-- contents -->
<!-- HTML footer for doxygen 1.8.9.1-->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.9.1
</small></address>
</body>
</html>
